// @x402/evm/dist/esm/chunk-DR77D5IU.mjs
import {
  trySignEip2612PermitExtension,
  trySignErc20ApprovalExtension
} from "./chunk-27MWX225.mjs";
import {
  PERMIT2_ADDRESS,
  uptoPermit2WitnessTypes,
  x402UptoPermit2ProxyAddress
} from "./chunk-MACPBXCT.mjs";
import {
  createPermit2Nonce,
  getEvmChainId
} from "./chunk-TW7Z65AO.mjs";

// src/upto/client/permit2.ts
import { getAddress } from "viem";
async function createUptoPermit2Payload(signer, x402Version, paymentRequirements) {
  const facilitatorAddress = paymentRequirements.extra?.facilitatorAddress;
  if (!facilitatorAddress) {
    throw new Error(
      "upto scheme requires facilitatorAddress in paymentRequirements.extra. Ensure the server is configured with an upto facilitator that provides getExtra()."
    );
  }
  const now = Math.floor(Date.now() / 1e3);
  const nonce = createPermit2Nonce();
  const validAfter = "0";
  const deadline = (now + paymentRequirements.maxTimeoutSeconds).toString();
  if (BigInt(deadline) <= BigInt(validAfter)) {
    throw new Error(
      `Invalid time window: deadline (${deadline}) must be after validAfter (${validAfter}). Check that maxTimeoutSeconds (${paymentRequirements.maxTimeoutSeconds}) is positive.`
    );
  }
  const permit2Authorization = {
    from: signer.address,
    permitted: {
      token: getAddress(paymentRequirements.asset),
      amount: paymentRequirements.amount
    },
    spender: x402UptoPermit2ProxyAddress,
    nonce,
    deadline,
    witness: {
      to: getAddress(paymentRequirements.payTo),
      facilitator: getAddress(facilitatorAddress),
      validAfter
    }
  };
  const chainId = getEvmChainId(paymentRequirements.network);
  const signature = await signer.signTypedData({
    domain: { name: "Permit2", chainId, verifyingContract: PERMIT2_ADDRESS },
    types: uptoPermit2WitnessTypes,
    primaryType: "PermitWitnessTransferFrom",
    message: {
      permitted: {
        token: getAddress(permit2Authorization.permitted.token),
        amount: BigInt(permit2Authorization.permitted.amount)
      },
      spender: getAddress(permit2Authorization.spender),
      nonce: BigInt(permit2Authorization.nonce),
      deadline: BigInt(permit2Authorization.deadline),
      witness: {
        to: getAddress(permit2Authorization.witness.to),
        facilitator: getAddress(permit2Authorization.witness.facilitator),
        validAfter: BigInt(permit2Authorization.witness.validAfter)
      }
    }
  });
  return {
    x402Version,
    payload: { signature, permit2Authorization }
  };
}

// src/upto/client/scheme.ts
var UptoEvmScheme = class {
  /**
   * Creates a new UptoEvmScheme instance.
   *
   * @param signer - The EVM signer for client operations
   * @param options - Optional RPC configuration
   */
  constructor(signer, options) {
    this.signer = signer;
    this.options = options;
    this.scheme = "upto";
  }
  /**
   * Creates a payment payload for the Upto scheme using Permit2.
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @param context - Optional context with server-declared extensions
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements, context) {
    const result = await createUptoPermit2Payload(this.signer, x402Version, paymentRequirements);
    const eip2612Extensions = await trySignEip2612PermitExtension(
      this.signer,
      this.options,
      paymentRequirements,
      result,
      context
    );
    if (eip2612Extensions) {
      return { ...result, extensions: eip2612Extensions };
    }
    const erc20Extensions = await trySignErc20ApprovalExtension(
      this.signer,
      this.options,
      paymentRequirements,
      context
    );
    if (erc20Extensions) {
      return { ...result, extensions: erc20Extensions };
    }
    return result;
  }
};

export {
  UptoEvmScheme
};
