// @x402/evm/dist/esm/chunk-LZBYVN3C.mjs
import {
  ExactEvmSchemeV12 as ExactEvmSchemeV1,
  NETWORKS
} from "./chunk-XYM7YNDI.mjs";
import {
  createPermit2Payload
} from "./chunk-GJOZC3AC.mjs";
import {
  trySignEip2612PermitExtension,
  trySignErc20ApprovalExtension
} from "./chunk-27MWX225.mjs";
import {
  authorizationTypes
} from "./chunk-MACPBXCT.mjs";
import {
  createNonce,
  getEvmChainId
} from "./chunk-TW7Z65AO.mjs";

// src/exact/client/eip3009.ts
import { getAddress } from "viem";
async function createEIP3009Payload(signer, x402Version, paymentRequirements) {
  const nonce = createNonce();
  const now = Math.floor(Date.now() / 1e3);
  const authorization = {
    from: signer.address,
    to: getAddress(paymentRequirements.payTo),
    value: paymentRequirements.amount,
    validAfter: "0",
    validBefore: (now + paymentRequirements.maxTimeoutSeconds).toString(),
    nonce
  };
  const signature = await signEIP3009Authorization(signer, authorization, paymentRequirements);
  const payload = {
    authorization,
    signature
  };
  return {
    x402Version,
    payload
  };
}
async function signEIP3009Authorization(signer, authorization, requirements) {
  const chainId = getEvmChainId(requirements.network);
  if (!requirements.extra?.name || !requirements.extra?.version) {
    throw new Error(
      `EIP-712 domain parameters (name, version) are required in payment requirements for asset ${requirements.asset}`
    );
  }
  const { name, version } = requirements.extra;
  const domain = {
    name,
    version,
    chainId,
    verifyingContract: getAddress(requirements.asset)
  };
  const message = {
    from: getAddress(authorization.from),
    to: getAddress(authorization.to),
    value: BigInt(authorization.value),
    validAfter: BigInt(authorization.validAfter),
    validBefore: BigInt(authorization.validBefore),
    nonce: authorization.nonce
  };
  return await signer.signTypedData({
    domain,
    types: authorizationTypes,
    primaryType: "TransferWithAuthorization",
    message
  });
}

// src/exact/client/scheme.ts
var ExactEvmScheme = class {
  /**
   * Creates a new ExactEvmClient instance.
   *
   * @param signer - The EVM signer for client operations.
   *   Base flow only requires `address` + `signTypedData`.
   *   Extension enrichment (EIP-2612 / ERC-20 approval sponsoring) additionally
   *   requires optional capabilities like `readContract` and tx signing helpers.
   * @param options - Optional RPC configuration used to backfill extension capabilities.
   */
  constructor(signer, options) {
    this.signer = signer;
    this.options = options;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   * Routes to EIP-3009 or Permit2 based on requirements.extra.assetTransferMethod.
   *
   * For Permit2 flows, if the server advertises `eip2612GasSponsoring` and the
   * signer supports `readContract`, automatically signs an EIP-2612 permit
   * when Permit2 allowance is insufficient.
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @param context - Optional context with server-declared extensions
   * @returns Promise resolving to a payment payload result (with optional extensions)
   */
  async createPaymentPayload(x402Version, paymentRequirements, context) {
    const assetTransferMethod = paymentRequirements.extra?.assetTransferMethod ?? "eip3009";
    if (assetTransferMethod === "permit2") {
      const result = await createPermit2Payload(this.signer, x402Version, paymentRequirements);
      const eip2612Extensions = await trySignEip2612PermitExtension(
        this.signer,
        this.options,
        paymentRequirements,
        result,
        context
      );
      if (eip2612Extensions) {
        return {
          ...result,
          extensions: eip2612Extensions
        };
      }
      const erc20Extensions = await trySignErc20ApprovalExtension(
        this.signer,
        this.options,
        paymentRequirements,
        context
      );
      if (erc20Extensions) {
        return {
          ...result,
          extensions: erc20Extensions
        };
      }
      return result;
    }
    return createEIP3009Payload(this.signer, x402Version, paymentRequirements);
  }
};

// src/exact/client/register.ts
function registerExactEvmScheme(client, config) {
  const evmScheme = new ExactEvmScheme(config.signer, config.schemeOptions);
  if (config.networks && config.networks.length > 0) {
    config.networks.forEach((network) => {
      client.register(network, evmScheme);
    });
  } else {
    client.register("eip155:*", evmScheme);
  }
  NETWORKS.forEach((network) => {
    client.registerV1(network, new ExactEvmSchemeV1(config.signer));
  });
  if (config.policies) {
    config.policies.forEach((policy) => {
      client.registerPolicy(policy);
    });
  }
  return client;
}

export {
  ExactEvmScheme,
  registerExactEvmScheme
};
