// @x402/evm/dist/esm/index.mjs
import {
  AUTH_CAPTURE_ESCROW_ADDRESS,
  AUTH_CAPTURE_SCHEME,
  AuthCaptureEvmScheme,
  EIP3009_TOKEN_COLLECTOR_ADDRESS,
  PERMIT2_TOKEN_COLLECTOR_ADDRESS
} from "./chunk-7KTOBWB2.mjs";
import {
  BatchSettlementEvmScheme
} from "./chunk-EXF5MSQO.mjs";
import "./chunk-W6ON4LG2.mjs";
import {
  BATCH_SETTLEMENT_ADDRESS,
  BATCH_SETTLEMENT_DOMAIN,
  BATCH_SETTLEMENT_SCHEME,
  ERC3009_DEPOSIT_COLLECTOR_ADDRESS,
  claimBatchTypes,
  refundTypes,
  voucherTypes
} from "./chunk-HYABYUBD.mjs";
import {
  ExactEvmScheme
} from "./chunk-LZBYVN3C.mjs";
import {
  DEFAULT_STABLECOINS,
  getDefaultAsset
} from "./chunk-VQXOENVL.mjs";
import {
  isEIP3009Payload,
  isPermit2Payload,
  isUptoPermit2Payload
} from "./chunk-JK7SLLF7.mjs";
import {
  isBatchSettlementClaimPayload,
  isBatchSettlementDepositPayload,
  isBatchSettlementEnrichedRefundPayload,
  isBatchSettlementRefundPayload,
  isBatchSettlementSettlePayload,
  isBatchSettlementVoucherPayload
} from "./chunk-U4HCGTLU.mjs";
import "./chunk-XYM7YNDI.mjs";
import {
  UptoEvmScheme
} from "./chunk-DR77D5IU.mjs";
import {
  createPermit2ApprovalTx,
  getPermit2AllowanceReadParams
} from "./chunk-GJOZC3AC.mjs";
import "./chunk-CE5P6FKO.mjs";
import "./chunk-P3QOX3QZ.mjs";
import {
  BUILDER_CODE_KEY,
  appendDataSuffix,
  resolveDataSuffix
} from "./chunk-27MWX225.mjs";
import {
  PERMIT2_ADDRESS,
  authorizationTypes,
  eip3009ABI,
  erc20AllowanceAbi,
  permit2WitnessTypes,
  uptoPermit2WitnessTypes,
  x402ExactPermit2ProxyABI,
  x402ExactPermit2ProxyAddress,
  x402UptoPermit2ProxyABI,
  x402UptoPermit2ProxyAddress
} from "./chunk-MACPBXCT.mjs";
import "./chunk-VS3RYAYE.mjs";
import "./chunk-TW7Z65AO.mjs";

// src/signer.ts
function toClientEvmSigner(signer, publicClient) {
  const readContract = signer.readContract ?? publicClient?.readContract.bind(publicClient);
  const result = {
    address: signer.address,
    signTypedData: (msg) => signer.signTypedData(msg)
  };
  if (readContract) {
    result.readContract = readContract;
  }
  const signTransaction = signer.signTransaction;
  if (signTransaction) {
    result.signTransaction = (args) => signTransaction(args);
  }
  const getTransactionCount = signer.getTransactionCount ?? publicClient?.getTransactionCount?.bind(publicClient);
  if (getTransactionCount) {
    result.getTransactionCount = (args) => getTransactionCount(args);
  }
  const estimateFeesPerGas = signer.estimateFeesPerGas ?? publicClient?.estimateFeesPerGas?.bind(publicClient);
  if (estimateFeesPerGas) {
    result.estimateFeesPerGas = () => estimateFeesPerGas();
  }
  return result;
}
function toFacilitatorEvmSigner(client) {
  return {
    ...client,
    getAddresses: () => [client.address]
  };
}

// src/auth-capture/types.ts
function isAuthCaptureExtra(value) {
  if (typeof value !== "object" || value === null) return false;
  const v = value;
  return typeof v.captureAuthorizer === "string" && typeof v.captureDeadline === "number" && typeof v.refundDeadline === "number" && typeof v.feeRecipient === "string" && typeof v.minFeeBps === "number" && typeof v.maxFeeBps === "number" && typeof v.name === "string" && typeof v.version === "string";
}
function isEip3009Payload(value) {
  if (typeof value !== "object" || value === null) return false;
  const v = value;
  return "authorization" in v && typeof v.authorization === "object" && v.authorization !== null && typeof v.signature === "string" && typeof v.salt === "string";
}
function isPermit2Payload2(value) {
  if (typeof value !== "object" || value === null) return false;
  const v = value;
  if (typeof v.signature !== "string" || typeof v.salt !== "string") return false;
  if (typeof v.permit2Authorization !== "object" || v.permit2Authorization === null) return false;
  const a = v.permit2Authorization;
  return typeof a.from === "string" && typeof a.spender === "string" && typeof a.nonce === "string" && typeof a.deadline === "string" && typeof a.permitted === "object" && a.permitted !== null;
}
function isAuthCapturePayload(value) {
  return isEip3009Payload(value) || isPermit2Payload2(value);
}
export {
  AUTH_CAPTURE_ESCROW_ADDRESS,
  AUTH_CAPTURE_SCHEME,
  AuthCaptureEvmScheme,
  BATCH_SETTLEMENT_ADDRESS,
  BATCH_SETTLEMENT_DOMAIN,
  BATCH_SETTLEMENT_SCHEME,
  BUILDER_CODE_KEY,
  BatchSettlementEvmScheme,
  DEFAULT_STABLECOINS,
  EIP3009_TOKEN_COLLECTOR_ADDRESS,
  ERC3009_DEPOSIT_COLLECTOR_ADDRESS,
  ExactEvmScheme,
  PERMIT2_ADDRESS,
  PERMIT2_TOKEN_COLLECTOR_ADDRESS,
  UptoEvmScheme,
  appendDataSuffix,
  authorizationTypes,
  claimBatchTypes,
  createPermit2ApprovalTx,
  eip3009ABI,
  erc20AllowanceAbi,
  getDefaultAsset,
  getPermit2AllowanceReadParams,
  isAuthCaptureExtra,
  isAuthCapturePayload,
  isBatchSettlementClaimPayload,
  isBatchSettlementDepositPayload,
  isBatchSettlementEnrichedRefundPayload,
  isBatchSettlementRefundPayload,
  isBatchSettlementSettlePayload,
  isBatchSettlementVoucherPayload,
  isEIP3009Payload,
  isPermit2Payload,
  isUptoPermit2Payload,
  permit2WitnessTypes,
  refundTypes,
  resolveDataSuffix,
  toClientEvmSigner,
  toFacilitatorEvmSigner,
  uptoPermit2WitnessTypes,
  voucherTypes,
  x402ExactPermit2ProxyABI,
  x402ExactPermit2ProxyAddress,
  x402UptoPermit2ProxyABI,
  x402UptoPermit2ProxyAddress
};
