// @x402/evm/dist/cjs/exact/facilitator/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/facilitator/index.ts
var facilitator_exports = {};
__export(facilitator_exports, {
  ExactEvmScheme: () => ExactEvmScheme,
  registerExactEvmScheme: () => registerExactEvmScheme
});
module.exports = __toCommonJS(facilitator_exports);

// src/types.ts
function isPermit2Payload(payload) {
  return "permit2Authorization" in payload;
}

// src/exact/facilitator/eip3009.ts
var import_viem8 = require("viem");

// src/constants.ts
var authorizationTypes = {
  TransferWithAuthorization: [
    { name: "from", type: "address" },
    { name: "to", type: "address" },
    { name: "value", type: "uint256" },
    { name: "validAfter", type: "uint256" },
    { name: "validBefore", type: "uint256" },
    { name: "nonce", type: "bytes32" }
  ]
};
var permit2WitnessTypes = {
  PermitWitnessTransferFrom: [
    { name: "permitted", type: "TokenPermissions" },
    { name: "spender", type: "address" },
    { name: "nonce", type: "uint256" },
    { name: "deadline", type: "uint256" },
    { name: "witness", type: "Witness" }
  ],
  TokenPermissions: [
    { name: "token", type: "address" },
    { name: "amount", type: "uint256" }
  ],
  Witness: [
    { name: "to", type: "address" },
    { name: "validAfter", type: "uint256" }
  ]
};
var eip3009ABI = [
  {
    inputs: [
      { name: "from", type: "address" },
      { name: "to", type: "address" },
      { name: "value", type: "uint256" },
      { name: "validAfter", type: "uint256" },
      { name: "validBefore", type: "uint256" },
      { name: "nonce", type: "bytes32" },
      { name: "v", type: "uint8" },
      { name: "r", type: "bytes32" },
      { name: "s", type: "bytes32" }
    ],
    name: "transferWithAuthorization",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function"
  },
  {
    inputs: [
      { name: "from", type: "address" },
      { name: "to", type: "address" },
      { name: "value", type: "uint256" },
      { name: "validAfter", type: "uint256" },
      { name: "validBefore", type: "uint256" },
      { name: "nonce", type: "bytes32" },
      { name: "signature", type: "bytes" }
    ],
    name: "transferWithAuthorization",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function"
  },
  {
    inputs: [{ name: "account", type: "address" }],
    name: "balanceOf",
    outputs: [{ name: "", type: "uint256" }],
    stateMutability: "view",
    type: "function"
  },
  {
    inputs: [],
    name: "version",
    outputs: [{ name: "", type: "string" }],
    stateMutability: "view",
    type: "function"
  },
  {
    inputs: [],
    name: "name",
    outputs: [{ name: "", type: "string" }],
    stateMutability: "view",
    type: "function"
  },
  {
    inputs: [
      { name: "authorizer", type: "address" },
      { name: "nonce", type: "bytes32" }
    ],
    name: "authorizationState",
    outputs: [{ name: "", type: "bool" }],
    stateMutability: "view",
    type: "function"
  }
];
var erc20ApproveAbi = [
  {
    type: "function",
    name: "approve",
    inputs: [
      { name: "spender", type: "address" },
      { name: "amount", type: "uint256" }
    ],
    outputs: [{ type: "bool" }],
    stateMutability: "nonpayable"
  }
];
var erc20AllowanceAbi = [
  {
    type: "function",
    name: "allowance",
    inputs: [
      { name: "owner", type: "address" },
      { name: "spender", type: "address" }
    ],
    outputs: [{ type: "uint256" }],
    stateMutability: "view"
  }
];
var PERMIT2_ADDRESS = "0x000000000022D473030F116dDEE9F6B43aC78BA3";
var x402ExactPermit2ProxyAddress = "0x402085c248EeA27D92E8b30b2C58ed07f9E20001";
var permit2WitnessABIComponents = [
  { name: "to", type: "address", internalType: "address" },
  { name: "validAfter", type: "uint256", internalType: "uint256" }
];
var x402ExactPermit2ProxyABI = [
  {
    type: "function",
    name: "PERMIT2",
    inputs: [],
    outputs: [{ name: "", type: "address", internalType: "contract ISignatureTransfer" }],
    stateMutability: "view"
  },
  {
    type: "function",
    name: "WITNESS_TYPEHASH",
    inputs: [],
    outputs: [{ name: "", type: "bytes32", internalType: "bytes32" }],
    stateMutability: "view"
  },
  {
    type: "function",
    name: "WITNESS_TYPE_STRING",
    inputs: [],
    outputs: [{ name: "", type: "string", internalType: "string" }],
    stateMutability: "view"
  },
  {
    type: "function",
    name: "settle",
    inputs: [
      {
        name: "permit",
        type: "tuple",
        internalType: "struct ISignatureTransfer.PermitTransferFrom",
        components: [
          {
            name: "permitted",
            type: "tuple",
            internalType: "struct ISignatureTransfer.TokenPermissions",
            components: [
              { name: "token", type: "address", internalType: "address" },
              { name: "amount", type: "uint256", internalType: "uint256" }
            ]
          },
          { name: "nonce", type: "uint256", internalType: "uint256" },
          { name: "deadline", type: "uint256", internalType: "uint256" }
        ]
      },
      { name: "owner", type: "address", internalType: "address" },
      {
        name: "witness",
        type: "tuple",
        internalType: "struct x402ExactPermit2Proxy.Witness",
        components: permit2WitnessABIComponents
      },
      { name: "signature", type: "bytes", internalType: "bytes" }
    ],
    outputs: [],
    stateMutability: "nonpayable"
  },
  {
    type: "function",
    name: "settleWithPermit",
    inputs: [
      {
        name: "permit2612",
        type: "tuple",
        internalType: "struct x402ExactPermit2Proxy.EIP2612Permit",
        components: [
          { name: "value", type: "uint256", internalType: "uint256" },
          { name: "deadline", type: "uint256", internalType: "uint256" },
          { name: "r", type: "bytes32", internalType: "bytes32" },
          { name: "s", type: "bytes32", internalType: "bytes32" },
          { name: "v", type: "uint8", internalType: "uint8" }
        ]
      },
      {
        name: "permit",
        type: "tuple",
        internalType: "struct ISignatureTransfer.PermitTransferFrom",
        components: [
          {
            name: "permitted",
            type: "tuple",
            internalType: "struct ISignatureTransfer.TokenPermissions",
            components: [
              { name: "token", type: "address", internalType: "address" },
              { name: "amount", type: "uint256", internalType: "uint256" }
            ]
          },
          { name: "nonce", type: "uint256", internalType: "uint256" },
          { name: "deadline", type: "uint256", internalType: "uint256" }
        ]
      },
      { name: "owner", type: "address", internalType: "address" },
      {
        name: "witness",
        type: "tuple",
        internalType: "struct x402ExactPermit2Proxy.Witness",
        components: permit2WitnessABIComponents
      },
      { name: "signature", type: "bytes", internalType: "bytes" }
    ],
    outputs: [],
    stateMutability: "nonpayable"
  },
  { type: "event", name: "Settled", inputs: [], anonymous: false },
  { type: "event", name: "SettledWithPermit", inputs: [], anonymous: false },
  { type: "error", name: "InvalidAmount", inputs: [] },
  { type: "error", name: "InvalidDestination", inputs: [] },
  { type: "error", name: "InvalidOwner", inputs: [] },
  { type: "error", name: "InvalidPermit2Address", inputs: [] },
  { type: "error", name: "PaymentTooEarly", inputs: [] },
  { type: "error", name: "Permit2612AmountMismatch", inputs: [] },
  { type: "error", name: "ReentrancyGuardReentrantCall", inputs: [] }
];

// src/utils.ts
var import_viem = require("viem");
function getEvmChainId(network) {
  if (network.startsWith("eip155:")) {
    const idStr = network.split(":")[1];
    const chainId = parseInt(idStr, 10);
    if (isNaN(chainId)) {
      throw new Error(`Invalid CAIP-2 chain ID: ${network}`);
    }
    return chainId;
  }
  throw new Error(`Unsupported network format: ${network} (expected eip155:CHAIN_ID)`);
}

// src/exact/facilitator/errors.ts
var ErrAssetNotDeployedContract = "asset_not_deployed_contract";
var ErrInvalidScheme = "invalid_exact_evm_scheme";
var ErrNetworkMismatch = "invalid_exact_evm_network_mismatch";
var ErrMissingEip712Domain = "invalid_exact_evm_missing_eip712_domain";
var ErrRecipientMismatch = "invalid_exact_evm_recipient_mismatch";
var ErrInvalidSignature = "invalid_exact_evm_signature";
var ErrValidBeforeExpired = "invalid_exact_evm_payload_authorization_valid_before";
var ErrValidAfterInFuture = "invalid_exact_evm_payload_authorization_valid_after";
var ErrInvalidAuthorizationValue = "invalid_exact_evm_authorization_value";
var ErrUndeployedSmartWallet = "invalid_exact_evm_payload_undeployed_smart_wallet";
var ErrTransactionFailed = "invalid_exact_evm_transaction_failed";
var ErrEip3009TokenNameMismatch = "invalid_exact_evm_token_name_mismatch";
var ErrEip3009TokenVersionMismatch = "invalid_exact_evm_token_version_mismatch";
var ErrEip3009NotSupported = "invalid_exact_evm_eip3009_not_supported";
var ErrEip3009NonceAlreadyUsed = "invalid_exact_evm_nonce_already_used";
var ErrEip3009InsufficientBalance = "invalid_exact_evm_insufficient_balance";
var ErrEip3009SimulationFailed = "invalid_exact_evm_transaction_simulation_failed";
var ErrPermit2InvalidSpender = "invalid_permit2_spender";
var ErrPermit2RecipientMismatch = "invalid_permit2_recipient_mismatch";
var ErrPermit2DeadlineExpired = "permit2_deadline_expired";
var ErrPermit2NotYetValid = "permit2_not_yet_valid";
var ErrPermit2AmountMismatch = "permit2_amount_mismatch";
var ErrPermit2TokenMismatch = "permit2_token_mismatch";
var ErrPermit2InvalidSignature = "invalid_permit2_signature";
var ErrPermit2AllowanceRequired = "permit2_allowance_required";
var ErrPermit2SimulationFailed = "permit2_simulation_failed";
var ErrPermit2InsufficientBalance = "permit2_insufficient_balance";
var ErrPermit2ProxyNotDeployed = "permit2_proxy_not_deployed";
var ErrPermit2InvalidAmount = "permit2_invalid_amount";
var ErrPermit2InvalidDestination = "permit2_invalid_destination";
var ErrPermit2InvalidOwner = "permit2_invalid_owner";
var ErrPermit2PaymentTooEarly = "permit2_payment_too_early";
var ErrPermit2InvalidNonce = "permit2_invalid_nonce";
var ErrPermit2612AmountMismatch = "permit2_2612_amount_mismatch";
var ErrErc20ApprovalInvalidFormat = "invalid_erc20_approval_extension_format";
var ErrErc20ApprovalFromMismatch = "erc20_approval_from_mismatch";
var ErrErc20ApprovalAssetMismatch = "erc20_approval_asset_mismatch";
var ErrErc20ApprovalSpenderNotPermit2 = "erc20_approval_spender_not_permit2";
var ErrErc20ApprovalTxWrongTarget = "erc20_approval_tx_wrong_target";
var ErrErc20ApprovalTxWrongSelector = "erc20_approval_tx_wrong_selector";
var ErrErc20ApprovalTxWrongSpender = "erc20_approval_tx_wrong_spender";
var ErrErc20ApprovalTxInvalidCalldata = "erc20_approval_tx_invalid_calldata";
var ErrErc20ApprovalTxSignerMismatch = "erc20_approval_tx_signer_mismatch";
var ErrErc20ApprovalTxInvalidSignature = "erc20_approval_tx_invalid_signature";
var ErrErc20ApprovalTxParseFailed = "erc20_approval_tx_parse_failed";
var ErrErc20ApprovalTxFailed = "erc20_approval_tx_failed";
var ErrInvalidEip2612ExtensionFormat = "invalid_eip2612_extension_format";
var ErrEip2612FromMismatch = "eip2612_from_mismatch";
var ErrEip2612AssetMismatch = "eip2612_asset_mismatch";
var ErrEip2612SpenderNotPermit2 = "eip2612_spender_not_permit2";
var ErrEip2612DeadlineExpired = "eip2612_deadline_expired";
var ErrUnsupportedPayloadType = "unsupported_payload_type";
var ErrInvalidTransactionState = "invalid_transaction_state";
var ErrFactoryNotAllowed = "eip6492_factory_not_allowed";

// src/exact/extensions.ts
var EIP2612_GAS_SPONSORING_KEY = "eip2612GasSponsoring";
var ERC20_APPROVAL_GAS_SPONSORING_KEY = "erc20ApprovalGasSponsoring";
function _extractInfo(payload, extensionKey) {
  const extensions = payload.extensions;
  if (!extensions) return null;
  const extension = extensions[extensionKey];
  if (!extension?.info) return null;
  return extension.info;
}
function extractEip2612GasSponsoringInfo(payload) {
  const info = _extractInfo(payload, EIP2612_GAS_SPONSORING_KEY);
  if (!info) return null;
  if (!info.from || !info.asset || !info.spender || !info.amount || !info.nonce || !info.deadline || !info.signature || !info.version) {
    return null;
  }
  return info;
}
function validateEip2612GasSponsoringInfo(info) {
  const addressPattern = /^0x[a-fA-F0-9]{40}$/;
  const numericPattern = /^[0-9]+$/;
  const hexPattern = /^0x[a-fA-F0-9]+$/;
  const versionPattern = /^[0-9]+(\.[0-9]+)*$/;
  return addressPattern.test(info.from) && addressPattern.test(info.asset) && addressPattern.test(info.spender) && numericPattern.test(info.amount) && numericPattern.test(info.nonce) && numericPattern.test(info.deadline) && hexPattern.test(info.signature) && versionPattern.test(info.version);
}
function extractErc20ApprovalGasSponsoringInfo(payload) {
  const info = _extractInfo(payload, ERC20_APPROVAL_GAS_SPONSORING_KEY);
  if (!info) return null;
  if (!info.from || !info.asset || !info.spender || !info.amount || !info.signedTransaction || !info.version) {
    return null;
  }
  return info;
}
function validateErc20ApprovalGasSponsoringInfo(info) {
  const addressPattern = /^0x[a-fA-F0-9]{40}$/;
  const numericPattern = /^[0-9]+$/;
  const hexPattern = /^0x[a-fA-F0-9]+$/;
  const versionPattern = /^[0-9]+(\.[0-9]+)*$/;
  return addressPattern.test(info.from) && addressPattern.test(info.asset) && addressPattern.test(info.spender) && numericPattern.test(info.amount) && hexPattern.test(info.signedTransaction) && versionPattern.test(info.version);
}
function resolveErc20ApprovalExtensionSigner(extension, network) {
  if (!extension) return void 0;
  return extension.signerForNetwork?.(network) ?? extension.signer;
}

// src/shared/extensions/gasSponsoring.ts
var import_viem5 = require("viem");

// src/exact/client/eip2612.ts
var import_viem2 = require("viem");

// src/exact/client/erc20approval.ts
var import_viem3 = require("viem");

// src/shared/rpc.ts
var import_viem4 = require("viem");

// src/shared/extensions/builderCode.ts
var BUILDER_CODE_KEY = "builder-code";
var BUILDER_CODE_RESOLVER = async (context, ctx) => {
  const ext = context.getExtension(BUILDER_CODE_KEY);
  if (!ext?.buildDataSuffix) {
    return void 0;
  }
  return ext.buildDataSuffix(ctx);
};
var DATA_SUFFIX_RESOLVERS = [BUILDER_CODE_RESOLVER];
async function resolveDataSuffix(context, ctx) {
  if (!context) {
    return void 0;
  }
  const parts = [];
  for (const resolver of DATA_SUFFIX_RESOLVERS) {
    const suffix = await resolver(context, ctx);
    if (suffix && suffix !== "0x" && suffix.length > 2) {
      parts.push(suffix);
    }
  }
  if (parts.length === 0) {
    return void 0;
  }
  if (parts.length === 1) {
    return parts[0];
  }
  return parts.reduce((acc, part, index) => {
    if (index === 0) {
      return part;
    }
    const stripped = part.startsWith("0x") ? part.slice(2) : part;
    return `${acc}${stripped}`;
  });
}
function appendDataSuffix(calldata, suffix) {
  if (!suffix || suffix === "0x" || suffix.length <= 2) {
    return calldata;
  }
  const suffixHex = suffix.startsWith("0x") ? suffix.slice(2) : suffix;
  return `${calldata}${suffixHex}`;
}

// src/exact/facilitator/eip3009-utils.ts
var import_viem7 = require("viem");

// src/multicall.ts
var import_viem6 = require("viem");
var MULTICALL3_ADDRESS = "0xcA11bde05977b3631167028862bE2a173976CA11";
var multicall3ABI = [
  {
    inputs: [
      { name: "requireSuccess", type: "bool" },
      {
        name: "calls",
        type: "tuple[]",
        components: [
          { name: "target", type: "address" },
          { name: "callData", type: "bytes" }
        ]
      }
    ],
    name: "tryAggregate",
    outputs: [
      {
        name: "returnData",
        type: "tuple[]",
        components: [
          { name: "success", type: "bool" },
          { name: "returnData", type: "bytes" }
        ]
      }
    ],
    stateMutability: "payable",
    type: "function"
  }
];
async function multicall(readContract, calls) {
  const aggregateCalls = calls.map((call) => {
    if ("callData" in call) {
      return { target: call.address, callData: call.callData };
    }
    const callData = (0, import_viem6.encodeFunctionData)({
      abi: call.abi,
      functionName: call.functionName,
      args: call.args
    });
    return { target: call.address, callData };
  });
  const rawResults = await readContract({
    address: MULTICALL3_ADDRESS,
    abi: multicall3ABI,
    functionName: "tryAggregate",
    args: [false, aggregateCalls]
  });
  return rawResults.map((raw, i) => {
    if (!raw.success) {
      return {
        status: "failure",
        error: new Error(`multicall: call reverted (returnData: ${raw.returnData})`)
      };
    }
    const call = calls[i];
    if ("callData" in call) {
      return { status: "success", result: void 0 };
    }
    try {
      const decoded = (0, import_viem6.decodeFunctionResult)({
        abi: call.abi,
        functionName: call.functionName,
        data: raw.returnData
      });
      return { status: "success", result: decoded };
    } catch (err) {
      return {
        status: "failure",
        error: err instanceof Error ? err : new Error(String(err))
      };
    }
  });
}

// src/exact/facilitator/eip3009-utils.ts
async function simulateEip3009Transfer(signer, erc20Address, payload, eip6492Deployment) {
  const auth = payload.authorization;
  const transferArgs = [
    (0, import_viem7.getAddress)(auth.from),
    (0, import_viem7.getAddress)(auth.to),
    BigInt(auth.value),
    BigInt(auth.validAfter),
    BigInt(auth.validBefore),
    auth.nonce
  ];
  if (eip6492Deployment) {
    const { signature: innerSignature } = (0, import_viem7.parseErc6492Signature)(payload.signature);
    const transferCalldata = (0, import_viem7.encodeFunctionData)({
      abi: eip3009ABI,
      functionName: "transferWithAuthorization",
      args: [...transferArgs, innerSignature]
    });
    try {
      const results = await multicall(signer.readContract.bind(signer), [
        {
          address: (0, import_viem7.getAddress)(eip6492Deployment.factoryAddress),
          callData: eip6492Deployment.factoryCalldata
        },
        {
          address: erc20Address,
          callData: transferCalldata
        }
      ]);
      return results[1]?.status === "success";
    } catch {
      return false;
    }
  }
  const sig = payload.signature;
  const sigLength = sig.startsWith("0x") ? sig.length - 2 : sig.length;
  const isECDSA = sigLength === 130;
  try {
    if (isECDSA) {
      const parsedSig = (0, import_viem7.parseSignature)(sig);
      await signer.readContract({
        address: erc20Address,
        abi: eip3009ABI,
        functionName: "transferWithAuthorization",
        args: [
          ...transferArgs,
          parsedSig.v ?? parsedSig.yParity,
          parsedSig.r,
          parsedSig.s
        ]
      });
    } else {
      await signer.readContract({
        address: erc20Address,
        abi: eip3009ABI,
        functionName: "transferWithAuthorization",
        args: [...transferArgs, sig]
      });
    }
    return true;
  } catch {
    return false;
  }
}
async function diagnoseEip3009SimulationFailure(signer, erc20Address, payload, requirements, amountRequired) {
  const payer = payload.authorization.from;
  const diagnosticCalls = [
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "balanceOf",
      args: [payload.authorization.from]
    },
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "name"
    },
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "version"
    },
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "authorizationState",
      args: [payload.authorization.from, payload.authorization.nonce]
    }
  ];
  try {
    const results = await multicall(signer.readContract.bind(signer), diagnosticCalls);
    const [balanceResult, nameResult, versionResult, authStateResult] = results;
    if (authStateResult.status === "failure") {
      return { isValid: false, invalidReason: ErrEip3009NotSupported, payer };
    }
    if (authStateResult.status === "success" && authStateResult.result === true) {
      return { isValid: false, invalidReason: ErrEip3009NonceAlreadyUsed, payer };
    }
    if (nameResult.status === "success" && requirements.extra?.name && nameResult.result !== requirements.extra.name) {
      return { isValid: false, invalidReason: ErrEip3009TokenNameMismatch, payer };
    }
    if (versionResult.status === "success" && requirements.extra?.version && versionResult.result !== requirements.extra.version) {
      return { isValid: false, invalidReason: ErrEip3009TokenVersionMismatch, payer };
    }
    if (balanceResult.status === "success") {
      const balance = balanceResult.result;
      if (balance < BigInt(amountRequired)) {
        return {
          isValid: false,
          invalidReason: ErrEip3009InsufficientBalance,
          payer
        };
      }
    }
  } catch {
  }
  return { isValid: false, invalidReason: ErrEip3009SimulationFailed, payer };
}
function parseEip3009TransferError(error) {
  const msg = error instanceof Error ? error.message : String(error);
  if (/authorization.*(expired|valid before)/i.test(msg) || /AuthorizationExpired/i.test(msg)) {
    return ErrValidBeforeExpired;
  }
  if (/authorization.*not.*valid|AuthorizationNotYetValid/i.test(msg)) {
    return ErrValidAfterInFuture;
  }
  if (/authorization.*used|AuthorizationAlreadyUsed|AuthorizationUsedOrCanceled/i.test(msg)) {
    return ErrEip3009NonceAlreadyUsed;
  }
  if (/transfer.*exceeds.*balance|insufficient.*balance|ERC20InsufficientBalance/i.test(msg)) {
    return ErrEip3009InsufficientBalance;
  }
  if (/invalid.*signature|SignerMismatch|InvalidSignatureV|InvalidSignatureS/i.test(msg)) {
    return ErrInvalidSignature;
  }
  return ErrTransactionFailed;
}
async function executeTransferWithAuthorization(signer, erc20Address, payload, dataSuffix) {
  const { signature } = (0, import_viem7.parseErc6492Signature)(payload.signature);
  const signatureLength = signature.startsWith("0x") ? signature.length - 2 : signature.length;
  const isECDSA = signatureLength === 130;
  const auth = payload.authorization;
  const baseArgs = [
    (0, import_viem7.getAddress)(auth.from),
    (0, import_viem7.getAddress)(auth.to),
    BigInt(auth.value),
    BigInt(auth.validAfter),
    BigInt(auth.validBefore),
    auth.nonce
  ];
  let signatureArgs;
  if (isECDSA) {
    const parsedSig = (0, import_viem7.parseSignature)(signature);
    signatureArgs = [
      parsedSig.v || parsedSig.yParity,
      parsedSig.r,
      parsedSig.s
    ];
  } else {
    signatureArgs = [signature];
  }
  return signer.writeContract({
    address: erc20Address,
    abi: eip3009ABI,
    functionName: "transferWithAuthorization",
    args: [...baseArgs, ...signatureArgs],
    dataSuffix
  });
}

// src/exact/facilitator/eip3009.ts
async function verifyEIP3009(signer, payload, requirements, eip3009Payload, options) {
  const payer = eip3009Payload.authorization.from;
  let eip6492Deployment;
  if (payload.accepted.scheme !== "exact" || requirements.scheme !== "exact") {
    return {
      isValid: false,
      invalidReason: ErrInvalidScheme,
      payer
    };
  }
  if (!requirements.extra?.name || !requirements.extra?.version) {
    return {
      isValid: false,
      invalidReason: ErrMissingEip712Domain,
      payer
    };
  }
  const { name, version } = requirements.extra;
  const erc20Address = (0, import_viem8.getAddress)(requirements.asset);
  if (payload.accepted.network !== requirements.network) {
    return {
      isValid: false,
      invalidReason: ErrNetworkMismatch,
      payer
    };
  }
  const permitTypedData = {
    types: authorizationTypes,
    primaryType: "TransferWithAuthorization",
    domain: {
      name,
      version,
      chainId: getEvmChainId(requirements.network),
      verifyingContract: erc20Address
    },
    message: {
      from: eip3009Payload.authorization.from,
      to: eip3009Payload.authorization.to,
      value: BigInt(eip3009Payload.authorization.value),
      validAfter: BigInt(eip3009Payload.authorization.validAfter),
      validBefore: BigInt(eip3009Payload.authorization.validBefore),
      nonce: eip3009Payload.authorization.nonce
    }
  };
  let isValid = false;
  try {
    isValid = await signer.verifyTypedData({
      address: eip3009Payload.authorization.from,
      ...permitTypedData,
      signature: eip3009Payload.signature
    });
  } catch {
    isValid = false;
  }
  const signature = eip3009Payload.signature;
  const sigLen = signature.startsWith("0x") ? signature.length - 2 : signature.length;
  const erc6492Data = (0, import_viem8.parseErc6492Signature)(signature);
  const hasDeploymentInfo = erc6492Data.address && erc6492Data.data && !(0, import_viem8.isAddressEqual)(erc6492Data.address, "0x0000000000000000000000000000000000000000");
  if (hasDeploymentInfo) {
    eip6492Deployment = {
      factoryAddress: erc6492Data.address,
      factoryCalldata: erc6492Data.data
    };
  }
  if (!isValid) {
    const isSmartWallet = sigLen > 130;
    if (!isSmartWallet) {
      return {
        isValid: false,
        invalidReason: ErrInvalidSignature,
        payer
      };
    }
    const bytecode = await signer.getCode({ address: payer });
    const isDeployed = bytecode && bytecode !== "0x";
    if (!isDeployed && !hasDeploymentInfo) {
      return {
        isValid: false,
        invalidReason: ErrUndeployedSmartWallet,
        payer
      };
    }
  }
  if ((0, import_viem8.getAddress)(eip3009Payload.authorization.to) !== (0, import_viem8.getAddress)(requirements.payTo)) {
    return {
      isValid: false,
      invalidReason: ErrRecipientMismatch,
      payer
    };
  }
  const now = Math.floor(Date.now() / 1e3);
  if (BigInt(eip3009Payload.authorization.validBefore) < BigInt(now + 6)) {
    return {
      isValid: false,
      invalidReason: ErrValidBeforeExpired,
      payer
    };
  }
  if (BigInt(eip3009Payload.authorization.validAfter) > BigInt(now)) {
    return {
      isValid: false,
      invalidReason: ErrValidAfterInFuture,
      payer
    };
  }
  if (BigInt(eip3009Payload.authorization.value) !== BigInt(requirements.amount)) {
    return {
      isValid: false,
      invalidReason: ErrInvalidAuthorizationValue,
      payer
    };
  }
  const assetBytecode = await signer.getCode({ address: erc20Address });
  if (!assetBytecode || assetBytecode === "0x") {
    return { isValid: false, invalidReason: ErrAssetNotDeployedContract, payer };
  }
  if (options?.simulate !== false) {
    const simulationSucceeded = await simulateEip3009Transfer(
      signer,
      erc20Address,
      eip3009Payload,
      eip6492Deployment
    );
    if (!simulationSucceeded) {
      return diagnoseEip3009SimulationFailure(
        signer,
        erc20Address,
        eip3009Payload,
        requirements,
        requirements.amount
      );
    }
  }
  return {
    isValid: true,
    invalidReason: void 0,
    payer
  };
}
async function settleEIP3009(signer, payload, requirements, eip3009Payload, config, context) {
  const payer = eip3009Payload.authorization.from;
  const valid = await verifyEIP3009(signer, payload, requirements, eip3009Payload, {
    simulate: config.simulateInSettle ?? false
  });
  if (!valid.isValid) {
    return {
      success: false,
      network: payload.accepted.network,
      transaction: "",
      errorReason: valid.invalidReason ?? ErrInvalidScheme,
      payer
    };
  }
  try {
    const { address: factoryAddress, data: factoryCalldata } = (0, import_viem8.parseErc6492Signature)(
      eip3009Payload.signature
    );
    if (factoryAddress && factoryCalldata && !(0, import_viem8.isAddressEqual)(factoryAddress, "0x0000000000000000000000000000000000000000")) {
      const bytecode = await signer.getCode({ address: payer });
      if (!bytecode || bytecode === "0x") {
        const normalizedFactory = factoryAddress.toLowerCase();
        const isAllowed = (config.eip6492AllowedFactories ?? []).some(
          (allowed) => allowed.toLowerCase() === normalizedFactory
        );
        if (!isAllowed) {
          return {
            success: false,
            errorReason: ErrFactoryNotAllowed,
            transaction: "",
            network: payload.accepted.network,
            payer
          };
        }
        const deployTx = await signer.sendTransaction({
          to: factoryAddress,
          data: factoryCalldata
        });
        await signer.waitForTransactionReceipt({ hash: deployTx });
      }
    }
    const dataSuffix = await resolveDataSuffix(context, {
      paymentPayload: payload,
      paymentRequirements: requirements
    });
    const tx = await executeTransferWithAuthorization(
      signer,
      (0, import_viem8.getAddress)(requirements.asset),
      eip3009Payload,
      dataSuffix
    );
    const receipt = await signer.waitForTransactionReceipt({ hash: tx });
    if (receipt.status !== "success") {
      return {
        success: false,
        errorReason: ErrTransactionFailed,
        transaction: tx,
        network: payload.accepted.network,
        payer
      };
    }
    return {
      success: true,
      transaction: tx,
      network: payload.accepted.network,
      payer
    };
  } catch (error) {
    return {
      success: false,
      errorReason: parseEip3009TransferError(error),
      transaction: "",
      network: payload.accepted.network,
      payer
    };
  }
}

// src/exact/facilitator/permit2.ts
var import_viem11 = require("viem");

// src/shared/erc20approval.ts
var import_viem9 = require("viem");
var APPROVE_SELECTOR = "0x095ea7b3";
async function validateErc20ApprovalForPayment(info, payer, tokenAddress) {
  if (!validateErc20ApprovalGasSponsoringInfo(info)) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalInvalidFormat,
      invalidMessage: "ERC-20 approval extension info failed schema validation"
    };
  }
  if ((0, import_viem9.getAddress)(info.from) !== (0, import_viem9.getAddress)(payer)) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalFromMismatch,
      invalidMessage: `Expected from=${payer}, got ${info.from}`
    };
  }
  if ((0, import_viem9.getAddress)(info.asset) !== tokenAddress) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalAssetMismatch,
      invalidMessage: `Expected asset=${tokenAddress}, got ${info.asset}`
    };
  }
  if ((0, import_viem9.getAddress)(info.spender) !== (0, import_viem9.getAddress)(PERMIT2_ADDRESS)) {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalSpenderNotPermit2,
      invalidMessage: `Expected spender=${PERMIT2_ADDRESS}, got ${info.spender}`
    };
  }
  try {
    const serializedTx = info.signedTransaction;
    const tx = (0, import_viem9.parseTransaction)(serializedTx);
    if (!tx.to || (0, import_viem9.getAddress)(tx.to) !== tokenAddress) {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxWrongTarget,
        invalidMessage: `Transaction targets ${tx.to ?? "null"}, expected ${tokenAddress}`
      };
    }
    const data = tx.data ?? "0x";
    if (!data.startsWith(APPROVE_SELECTOR)) {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxWrongSelector,
        invalidMessage: `Transaction calldata does not start with approve() selector ${APPROVE_SELECTOR}`
      };
    }
    try {
      const decoded = (0, import_viem9.decodeFunctionData)({
        abi: erc20ApproveAbi,
        data
      });
      const calldataSpender = (0, import_viem9.getAddress)(decoded.args[0]);
      if (calldataSpender !== (0, import_viem9.getAddress)(PERMIT2_ADDRESS)) {
        return {
          isValid: false,
          invalidReason: ErrErc20ApprovalTxWrongSpender,
          invalidMessage: `approve() spender is ${calldataSpender}, expected Permit2 ${PERMIT2_ADDRESS}`
        };
      }
    } catch {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxInvalidCalldata,
        invalidMessage: "Failed to decode approve() calldata from the signed transaction"
      };
    }
    try {
      const recoveredAddress = await (0, import_viem9.recoverTransactionAddress)({
        serializedTransaction: serializedTx
      });
      if ((0, import_viem9.getAddress)(recoveredAddress) !== (0, import_viem9.getAddress)(payer)) {
        return {
          isValid: false,
          invalidReason: ErrErc20ApprovalTxSignerMismatch,
          invalidMessage: `Transaction signed by ${recoveredAddress}, expected payer ${payer}`
        };
      }
    } catch {
      return {
        isValid: false,
        invalidReason: ErrErc20ApprovalTxInvalidSignature,
        invalidMessage: "Failed to recover signer from the signed transaction"
      };
    }
  } catch {
    return {
      isValid: false,
      invalidReason: ErrErc20ApprovalTxParseFailed,
      invalidMessage: "Failed to parse the signed transaction"
    };
  }
  return { isValid: true };
}

// src/shared/permit2.ts
var import_viem10 = require("viem");

// src/upto/facilitator/errors.ts
var ErrUptoAmountExceedsPermitted = "upto_amount_exceeds_permitted";
var ErrUptoUnauthorizedFacilitator = "upto_unauthorized_facilitator";

// src/shared/permit2.ts
async function waitAndReturnSettleResponse(signer, tx, payload, payer) {
  const receipt = await signer.waitForTransactionReceipt({ hash: tx });
  if (receipt.status !== "success") {
    return {
      success: false,
      errorReason: ErrInvalidTransactionState,
      transaction: tx,
      network: payload.accepted.network,
      payer
    };
  }
  return {
    success: true,
    transaction: tx,
    network: payload.accepted.network,
    payer
  };
}
function mapSettleError(error, payload, payer) {
  let errorReason = ErrTransactionFailed;
  if (error instanceof Error) {
    const message = error.message;
    if (message.includes("Permit2612AmountMismatch")) {
      errorReason = ErrPermit2612AmountMismatch;
    } else if (message.includes("InvalidAmount")) {
      errorReason = ErrPermit2InvalidAmount;
    } else if (message.includes("InvalidDestination")) {
      errorReason = ErrPermit2InvalidDestination;
    } else if (message.includes("InvalidOwner")) {
      errorReason = ErrPermit2InvalidOwner;
    } else if (message.includes("PaymentTooEarly")) {
      errorReason = ErrPermit2PaymentTooEarly;
    } else if (message.includes("InvalidSignature") || message.includes("SignatureExpired")) {
      errorReason = ErrPermit2InvalidSignature;
    } else if (message.includes("InvalidNonce")) {
      errorReason = ErrPermit2InvalidNonce;
    } else if (message.includes("erc20_approval_tx_failed")) {
      errorReason = ErrErc20ApprovalTxFailed;
    } else if (message.includes("AmountExceedsPermitted")) {
      errorReason = ErrUptoAmountExceedsPermitted;
    } else if (message.includes("UnauthorizedFacilitator")) {
      errorReason = ErrUptoUnauthorizedFacilitator;
    } else {
      errorReason = `${ErrTransactionFailed}: ${message.slice(0, 500)}`;
    }
  }
  return {
    success: false,
    errorReason,
    transaction: "",
    network: payload.accepted.network,
    payer
  };
}
function validateEip2612PermitForPayment(info, payer, tokenAddress) {
  if (!validateEip2612GasSponsoringInfo(info)) {
    return { isValid: false, invalidReason: ErrInvalidEip2612ExtensionFormat };
  }
  if ((0, import_viem10.getAddress)(info.from) !== (0, import_viem10.getAddress)(payer)) {
    return { isValid: false, invalidReason: ErrEip2612FromMismatch };
  }
  if ((0, import_viem10.getAddress)(info.asset) !== tokenAddress) {
    return { isValid: false, invalidReason: ErrEip2612AssetMismatch };
  }
  if ((0, import_viem10.getAddress)(info.spender) !== (0, import_viem10.getAddress)(PERMIT2_ADDRESS)) {
    return { isValid: false, invalidReason: ErrEip2612SpenderNotPermit2 };
  }
  const now = Math.floor(Date.now() / 1e3);
  if (BigInt(info.deadline) < BigInt(now + 6)) {
    return { isValid: false, invalidReason: ErrEip2612DeadlineExpired };
  }
  return { isValid: true };
}
async function simulatePermit2Settle(config, signer, settleArgs) {
  try {
    await signer.readContract({
      address: config.proxyAddress,
      abi: config.proxyABI,
      functionName: "settle",
      args: settleArgs
    });
    return true;
  } catch {
    return false;
  }
}
async function simulatePermit2SettleWithPermit(config, signer, settleArgs, eip2612Info) {
  try {
    const { v, r, s } = splitEip2612Signature(eip2612Info.signature);
    await signer.readContract({
      address: config.proxyAddress,
      abi: config.proxyABI,
      functionName: "settleWithPermit",
      args: [
        {
          value: BigInt(eip2612Info.amount),
          deadline: BigInt(eip2612Info.deadline),
          r,
          s,
          v
        },
        ...settleArgs
      ]
    });
    return true;
  } catch {
    return false;
  }
}
async function simulatePermit2SettleWithErc20Approval(config, extensionSigner, settleArgs, erc20Info) {
  if (!extensionSigner.simulateTransactions) {
    return false;
  }
  try {
    const settleData = (0, import_viem10.encodeFunctionData)({
      abi: config.proxyABI,
      functionName: "settle",
      args: settleArgs
    });
    return await extensionSigner.simulateTransactions([
      erc20Info.signedTransaction,
      { to: config.proxyAddress, data: settleData, gas: BigInt(3e5) }
    ]);
  } catch {
    return false;
  }
}
async function diagnosePermit2SimulationFailure(config, signer, tokenAddress, permit2Payload, amountRequired) {
  const payer = permit2Payload.permit2Authorization.from;
  const diagnosticCalls = [
    {
      address: config.proxyAddress,
      abi: config.proxyABI,
      functionName: "PERMIT2"
    },
    {
      address: tokenAddress,
      abi: eip3009ABI,
      functionName: "balanceOf",
      args: [payer]
    },
    {
      address: tokenAddress,
      abi: erc20AllowanceAbi,
      functionName: "allowance",
      args: [payer, PERMIT2_ADDRESS]
    }
  ];
  try {
    const results = await multicall(signer.readContract.bind(signer), diagnosticCalls);
    const [proxyResult, balanceResult, allowanceResult] = results;
    if (proxyResult.status === "failure") {
      return { isValid: false, invalidReason: ErrPermit2ProxyNotDeployed, payer };
    }
    if (balanceResult.status === "success") {
      const balance = balanceResult.result;
      if (balance < BigInt(amountRequired)) {
        return { isValid: false, invalidReason: ErrPermit2InsufficientBalance, payer };
      }
    }
    if (allowanceResult.status === "success") {
      const allowance = allowanceResult.result;
      if (allowance < BigInt(amountRequired)) {
        return { isValid: false, invalidReason: ErrPermit2AllowanceRequired, payer };
      }
    }
  } catch {
  }
  return { isValid: false, invalidReason: ErrPermit2SimulationFailed, payer };
}
async function checkPermit2Prerequisites(config, signer, tokenAddress, payer, amountRequired) {
  const diagnosticCalls = [
    {
      address: config.proxyAddress,
      abi: config.proxyABI,
      functionName: "PERMIT2"
    },
    {
      address: tokenAddress,
      abi: eip3009ABI,
      functionName: "balanceOf",
      args: [payer]
    }
  ];
  try {
    const results = await multicall(signer.readContract.bind(signer), diagnosticCalls);
    const [proxyResult, balanceResult] = results;
    if (proxyResult.status === "failure") {
      return { isValid: false, invalidReason: ErrPermit2ProxyNotDeployed, payer };
    }
    if (balanceResult.status === "success") {
      const balance = balanceResult.result;
      if (balance < BigInt(amountRequired)) {
        return { isValid: false, invalidReason: ErrPermit2InsufficientBalance, payer };
      }
    }
  } catch {
  }
  return { isValid: true, invalidReason: void 0, payer };
}
function buildExactPermit2SettleArgs(permit2Payload) {
  const { signature } = (0, import_viem10.parseErc6492Signature)(permit2Payload.signature);
  return [
    {
      permitted: {
        token: (0, import_viem10.getAddress)(permit2Payload.permit2Authorization.permitted.token),
        amount: BigInt(permit2Payload.permit2Authorization.permitted.amount)
      },
      nonce: BigInt(permit2Payload.permit2Authorization.nonce),
      deadline: BigInt(permit2Payload.permit2Authorization.deadline)
    },
    (0, import_viem10.getAddress)(permit2Payload.permit2Authorization.from),
    {
      to: (0, import_viem10.getAddress)(permit2Payload.permit2Authorization.witness.to),
      validAfter: BigInt(permit2Payload.permit2Authorization.witness.validAfter)
    },
    signature
  ];
}
function splitEip2612Signature(signature) {
  const sig = signature.startsWith("0x") ? signature.slice(2) : signature;
  if (sig.length !== 130) {
    throw new Error(
      `invalid EIP-2612 signature length: expected 65 bytes (130 hex chars), got ${sig.length / 2} bytes`
    );
  }
  const r = `0x${sig.slice(0, 64)}`;
  const s = `0x${sig.slice(64, 128)}`;
  const v = parseInt(sig.slice(128, 130), 16);
  return { v, r, s };
}

// src/exact/facilitator/permit2.ts
var exactProxyConfig = {
  proxyAddress: x402ExactPermit2ProxyAddress,
  proxyABI: x402ExactPermit2ProxyABI
};
async function verifyPermit2(signer, payload, requirements, permit2Payload, context, options) {
  const payer = permit2Payload.permit2Authorization.from;
  if (payload.accepted.scheme !== "exact" || requirements.scheme !== "exact") {
    return {
      isValid: false,
      invalidReason: ErrUnsupportedPayloadType,
      payer
    };
  }
  if (payload.accepted.network !== requirements.network) {
    return {
      isValid: false,
      invalidReason: ErrNetworkMismatch,
      payer
    };
  }
  const chainId = getEvmChainId(requirements.network);
  const tokenAddress = (0, import_viem11.getAddress)(requirements.asset);
  const assetBytecode = await signer.getCode({ address: tokenAddress });
  if (!assetBytecode || assetBytecode === "0x") {
    return { isValid: false, invalidReason: ErrAssetNotDeployedContract, payer };
  }
  if ((0, import_viem11.getAddress)(permit2Payload.permit2Authorization.spender) !== (0, import_viem11.getAddress)(x402ExactPermit2ProxyAddress)) {
    return {
      isValid: false,
      invalidReason: ErrPermit2InvalidSpender,
      payer
    };
  }
  if ((0, import_viem11.getAddress)(permit2Payload.permit2Authorization.witness.to) !== (0, import_viem11.getAddress)(requirements.payTo)) {
    return {
      isValid: false,
      invalidReason: ErrPermit2RecipientMismatch,
      payer
    };
  }
  const now = Math.floor(Date.now() / 1e3);
  if (BigInt(permit2Payload.permit2Authorization.deadline) < BigInt(now + 6)) {
    return {
      isValid: false,
      invalidReason: ErrPermit2DeadlineExpired,
      payer
    };
  }
  if (BigInt(permit2Payload.permit2Authorization.witness.validAfter) > BigInt(now)) {
    return {
      isValid: false,
      invalidReason: ErrPermit2NotYetValid,
      payer
    };
  }
  if (BigInt(permit2Payload.permit2Authorization.permitted.amount) !== BigInt(requirements.amount)) {
    return {
      isValid: false,
      invalidReason: ErrPermit2AmountMismatch,
      payer
    };
  }
  if ((0, import_viem11.getAddress)(permit2Payload.permit2Authorization.permitted.token) !== tokenAddress) {
    return {
      isValid: false,
      invalidReason: ErrPermit2TokenMismatch,
      payer
    };
  }
  const permit2TypedData = {
    types: permit2WitnessTypes,
    primaryType: "PermitWitnessTransferFrom",
    domain: {
      name: "Permit2",
      chainId,
      verifyingContract: PERMIT2_ADDRESS
    },
    message: {
      permitted: {
        token: (0, import_viem11.getAddress)(permit2Payload.permit2Authorization.permitted.token),
        amount: BigInt(permit2Payload.permit2Authorization.permitted.amount)
      },
      spender: (0, import_viem11.getAddress)(permit2Payload.permit2Authorization.spender),
      nonce: BigInt(permit2Payload.permit2Authorization.nonce),
      deadline: BigInt(permit2Payload.permit2Authorization.deadline),
      witness: {
        to: (0, import_viem11.getAddress)(permit2Payload.permit2Authorization.witness.to),
        validAfter: BigInt(permit2Payload.permit2Authorization.witness.validAfter)
      }
    }
  };
  let signatureValid = false;
  try {
    signatureValid = await signer.verifyTypedData({
      address: payer,
      ...permit2TypedData,
      signature: permit2Payload.signature
    });
  } catch {
    signatureValid = false;
  }
  if (!signatureValid) {
    const bytecode = await signer.getCode({ address: payer });
    const isDeployedContract = bytecode && bytecode !== "0x";
    if (!isDeployedContract) {
      return {
        isValid: false,
        invalidReason: ErrPermit2InvalidSignature,
        payer
      };
    }
  }
  if (options?.simulate === false) {
    return { isValid: true, invalidReason: void 0, payer };
  }
  const eip2612Info = extractEip2612GasSponsoringInfo(payload);
  if (eip2612Info) {
    const fieldResult = validateEip2612PermitForPayment(eip2612Info, payer, tokenAddress);
    if (!fieldResult.isValid) {
      return { isValid: false, invalidReason: fieldResult.invalidReason, payer };
    }
    const exactSettleArgs = buildExactPermit2SettleArgs(permit2Payload);
    const simOk2 = await simulatePermit2SettleWithPermit(
      exactProxyConfig,
      signer,
      exactSettleArgs,
      eip2612Info
    );
    if (!simOk2) {
      return diagnosePermit2SimulationFailure(
        exactProxyConfig,
        signer,
        tokenAddress,
        permit2Payload,
        requirements.amount
      );
    }
    return { isValid: true, invalidReason: void 0, payer };
  }
  const erc20GasSponsorshipExtension = context?.getExtension(
    ERC20_APPROVAL_GAS_SPONSORING_KEY
  );
  if (erc20GasSponsorshipExtension) {
    const erc20Info = extractErc20ApprovalGasSponsoringInfo(payload);
    if (erc20Info) {
      const fieldResult = await validateErc20ApprovalForPayment(erc20Info, payer, tokenAddress);
      if (!fieldResult.isValid) {
        return { isValid: false, invalidReason: fieldResult.invalidReason, payer };
      }
      const extensionSigner = resolveErc20ApprovalExtensionSigner(
        erc20GasSponsorshipExtension,
        requirements.network
      );
      if (extensionSigner?.simulateTransactions) {
        const simOk2 = await simulatePermit2SettleWithErc20Approval(
          exactProxyConfig,
          extensionSigner,
          buildExactPermit2SettleArgs(permit2Payload),
          erc20Info
        );
        if (!simOk2) {
          return diagnosePermit2SimulationFailure(
            exactProxyConfig,
            signer,
            tokenAddress,
            permit2Payload,
            requirements.amount
          );
        }
        return { isValid: true, invalidReason: void 0, payer };
      }
      return checkPermit2Prerequisites(
        exactProxyConfig,
        signer,
        tokenAddress,
        payer,
        requirements.amount
      );
    }
  }
  const simOk = await simulatePermit2Settle(
    exactProxyConfig,
    signer,
    buildExactPermit2SettleArgs(permit2Payload)
  );
  if (!simOk) {
    return diagnosePermit2SimulationFailure(
      exactProxyConfig,
      signer,
      tokenAddress,
      permit2Payload,
      requirements.amount
    );
  }
  return { isValid: true, invalidReason: void 0, payer };
}
async function settlePermit2(signer, payload, requirements, permit2Payload, context, config) {
  const payer = permit2Payload.permit2Authorization.from;
  const valid = await verifyPermit2(signer, payload, requirements, permit2Payload, context, {
    simulate: config?.simulateInSettle ?? false
  });
  if (!valid.isValid) {
    return {
      success: false,
      network: payload.accepted.network,
      transaction: "",
      errorReason: valid.invalidReason ?? ErrInvalidScheme,
      payer
    };
  }
  const dataSuffix = await resolveDataSuffix(context, {
    paymentPayload: payload,
    paymentRequirements: requirements
  });
  const eip2612Info = extractEip2612GasSponsoringInfo(payload);
  if (eip2612Info) {
    return settlePermit2WithEIP2612(
      exactProxyConfig,
      signer,
      payload,
      permit2Payload,
      eip2612Info,
      dataSuffix
    );
  }
  const erc20Info = extractErc20ApprovalGasSponsoringInfo(payload);
  if (erc20Info) {
    const erc20GasSponsorshipExtension = context?.getExtension(
      ERC20_APPROVAL_GAS_SPONSORING_KEY
    );
    const extensionSigner = resolveErc20ApprovalExtensionSigner(
      erc20GasSponsorshipExtension,
      payload.accepted.network
    );
    if (extensionSigner) {
      return settlePermit2WithERC20Approval(
        exactProxyConfig,
        extensionSigner,
        payload,
        permit2Payload,
        erc20Info,
        dataSuffix
      );
    }
  }
  return settlePermit2Direct(exactProxyConfig, signer, payload, permit2Payload, dataSuffix);
}
async function settlePermit2WithEIP2612(config, signer, payload, permit2Payload, eip2612Info, dataSuffix) {
  const payer = permit2Payload.permit2Authorization.from;
  try {
    const { v, r, s } = splitEip2612Signature(eip2612Info.signature);
    const tx = await signer.writeContract({
      address: config.proxyAddress,
      abi: config.proxyABI,
      functionName: "settleWithPermit",
      args: [
        {
          value: BigInt(eip2612Info.amount),
          deadline: BigInt(eip2612Info.deadline),
          r,
          s,
          v
        },
        ...buildExactPermit2SettleArgs(permit2Payload)
      ],
      dataSuffix
    });
    return waitAndReturnSettleResponse(signer, tx, payload, payer);
  } catch (error) {
    return mapSettleError(error, payload, payer);
  }
}
async function settlePermit2WithERC20Approval(config, extensionSigner, payload, permit2Payload, erc20Info, dataSuffix) {
  const payer = permit2Payload.permit2Authorization.from;
  try {
    const settleData = appendDataSuffix(
      (0, import_viem11.encodeFunctionData)({
        abi: config.proxyABI,
        functionName: "settle",
        args: buildExactPermit2SettleArgs(permit2Payload)
      }),
      dataSuffix
    );
    const txHashes = await extensionSigner.sendTransactions([
      erc20Info.signedTransaction,
      { to: config.proxyAddress, data: settleData, gas: BigInt(3e5) }
    ]);
    const settleTxHash = txHashes[txHashes.length - 1];
    return waitAndReturnSettleResponse(extensionSigner, settleTxHash, payload, payer);
  } catch (error) {
    return mapSettleError(error, payload, payer);
  }
}
async function settlePermit2Direct(config, signer, payload, permit2Payload, dataSuffix) {
  const payer = permit2Payload.permit2Authorization.from;
  try {
    const tx = await signer.writeContract({
      address: config.proxyAddress,
      abi: config.proxyABI,
      functionName: "settle",
      args: buildExactPermit2SettleArgs(permit2Payload),
      dataSuffix
    });
    return waitAndReturnSettleResponse(signer, tx, payload, payer);
  } catch (error) {
    return mapSettleError(error, payload, payer);
  }
}

// src/exact/facilitator/scheme.ts
var ExactEvmScheme = class {
  /**
   * Creates a new ExactEvmScheme facilitator instance.
   *
   * @param signer - The EVM signer for facilitator operations
   * @param config - Optional configuration
   */
  constructor(signer, config) {
    this.signer = signer;
    this.scheme = "exact";
    this.caipFamily = "eip155:*";
    this.config = {
      eip6492AllowedFactories: config?.eip6492AllowedFactories ?? [],
      simulateInSettle: config?.simulateInSettle ?? false
    };
  }
  /**
   * Returns undefined — EVM has no mechanism-specific extra data.
   *
   * @param _ - The network identifier (unused)
   * @returns undefined
   */
  getExtra(_) {
    return void 0;
  }
  /**
   * Returns facilitator wallet addresses for the supported response.
   *
   * @param _ - The network identifier (unused, addresses are network-agnostic)
   * @returns Array of facilitator wallet addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies a payment payload. Routes to Permit2 or EIP-3009 based on payload type.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @param context - Optional facilitator context for extension capabilities
   * @param _ - Payment required extensions (unused; reserved for interface parity)
   * @returns Promise resolving to verification response
   */
  async verify(payload, requirements, context, _) {
    const rawPayload = payload.payload;
    const isPermit2 = isPermit2Payload(rawPayload);
    if (isPermit2) {
      return verifyPermit2(this.signer, payload, requirements, rawPayload, context);
    }
    const eip3009Payload = rawPayload;
    return verifyEIP3009(this.signer, payload, requirements, eip3009Payload);
  }
  /**
   * Settles a payment. Routes to Permit2 or EIP-3009 based on payload type.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @param context - Optional facilitator context for extension capabilities
   * @returns Promise resolving to settlement response
   */
  async settle(payload, requirements, context) {
    const rawPayload = payload.payload;
    const isPermit2 = isPermit2Payload(rawPayload);
    if (isPermit2) {
      return settlePermit2(this.signer, payload, requirements, rawPayload, context, {
        simulateInSettle: this.config.simulateInSettle
      });
    }
    const eip3009Payload = rawPayload;
    return settleEIP3009(this.signer, payload, requirements, eip3009Payload, this.config, context);
  }
};

// src/exact/v1/facilitator/scheme.ts
var import_viem13 = require("viem");

// src/exact/v1/client/scheme.ts
var import_viem12 = require("viem");

// src/v1/index.ts
var EVM_NETWORK_CHAIN_ID_MAP = {
  ethereum: 1,
  sepolia: 11155111,
  abstract: 2741,
  "abstract-testnet": 11124,
  "base-sepolia": 84532,
  base: 8453,
  "avalanche-fuji": 43113,
  avalanche: 43114,
  iotex: 4689,
  sei: 1329,
  "sei-testnet": 1328,
  polygon: 137,
  "polygon-amoy": 80002,
  peaq: 3338,
  story: 1514,
  educhain: 41923,
  "skale-base-sepolia": 324705682,
  megaeth: 4326,
  monad: 143,
  stable: 988,
  "stable-testnet": 2201
};
var NETWORKS = Object.keys(EVM_NETWORK_CHAIN_ID_MAP);
function getEvmChainIdV1(network) {
  const chainId = EVM_NETWORK_CHAIN_ID_MAP[network];
  if (!chainId) {
    throw new Error(`Unsupported v1 network: ${network}`);
  }
  return chainId;
}

// src/exact/v1/facilitator/scheme.ts
var ExactEvmSchemeV12 = class {
  /**
   * Creates a new ExactEvmFacilitatorV1 instance.
   *
   * @param signer - The EVM signer for facilitator operations
   * @param config - Optional configuration for the facilitator
   */
  constructor(signer, config) {
    this.signer = signer;
    this.scheme = "exact";
    this.caipFamily = "eip155:*";
    this.config = {
      eip6492AllowedFactories: config?.eip6492AllowedFactories ?? [],
      simulateInSettle: config?.simulateInSettle ?? false
    };
  }
  /**
   * Get mechanism-specific extra data for the supported kinds endpoint.
   * For EVM, no extra data is needed.
   *
   * @param _ - The network identifier (unused for EVM)
   * @returns undefined (EVM has no extra data)
   */
  getExtra(_) {
    return void 0;
  }
  /**
   * Get signer addresses used by this facilitator.
   * Returns all addresses this facilitator can use for signing/settling transactions.
   *
   * @param _ - The network identifier (unused for EVM, addresses are network-agnostic)
   * @returns Array of facilitator wallet addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies a payment payload (V1).
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @returns Promise resolving to verification response
   */
  async verify(payload, requirements) {
    return this._verify(payload, requirements);
  }
  /**
   * Settles a payment by executing the transfer (V1).
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @param context - Optional facilitator context for extension capabilities
   * @returns Promise resolving to settlement response
   */
  async settle(payload, requirements, context) {
    const payloadV1 = payload;
    const exactEvmPayload = payload.payload;
    const valid = await this._verify(payload, requirements, {
      simulate: this.config.simulateInSettle ?? false
    });
    if (!valid.isValid) {
      return {
        success: false,
        network: payloadV1.network,
        transaction: "",
        errorReason: valid.invalidReason ?? ErrInvalidScheme,
        payer: exactEvmPayload.authorization.from
      };
    }
    try {
      const { address: factoryAddress, data: factoryCalldata } = (0, import_viem13.parseErc6492Signature)(
        exactEvmPayload.signature
      );
      if (factoryAddress && factoryCalldata && !(0, import_viem13.isAddressEqual)(factoryAddress, "0x0000000000000000000000000000000000000000")) {
        const payerAddress = exactEvmPayload.authorization.from;
        const bytecode = await this.signer.getCode({ address: payerAddress });
        if (!bytecode || bytecode === "0x") {
          const normalizedFactory = factoryAddress.toLowerCase();
          const isAllowed = (this.config.eip6492AllowedFactories ?? []).some(
            (allowed) => allowed.toLowerCase() === normalizedFactory
          );
          if (!isAllowed) {
            return {
              success: false,
              errorReason: ErrFactoryNotAllowed,
              transaction: "",
              network: payloadV1.network,
              payer: exactEvmPayload.authorization.from
            };
          }
          const deployTx = await this.signer.sendTransaction({
            to: factoryAddress,
            data: factoryCalldata
          });
          await this.signer.waitForTransactionReceipt({ hash: deployTx });
        }
      }
      const dataSuffix = await resolveDataSuffix(context, {
        paymentPayload: payload,
        paymentRequirements: requirements
      });
      const tx = await executeTransferWithAuthorization(
        this.signer,
        (0, import_viem13.getAddress)(requirements.asset),
        exactEvmPayload,
        dataSuffix
      );
      const receipt = await this.signer.waitForTransactionReceipt({ hash: tx });
      if (receipt.status !== "success") {
        return {
          success: false,
          errorReason: ErrTransactionFailed,
          transaction: tx,
          network: payloadV1.network,
          payer: exactEvmPayload.authorization.from
        };
      }
      return {
        success: true,
        transaction: tx,
        network: payloadV1.network,
        payer: exactEvmPayload.authorization.from
      };
    } catch (error) {
      return {
        success: false,
        errorReason: error instanceof Error ? error.message : ErrTransactionFailed,
        transaction: "",
        network: payloadV1.network,
        payer: exactEvmPayload.authorization.from
      };
    }
  }
  /**
   * Internal verify with optional simulation control.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @param options - Verification options (e.g. simulate)
   * @returns Promise resolving to verification response
   */
  async _verify(payload, requirements, options) {
    const requirementsV1 = requirements;
    const payloadV1 = payload;
    const exactEvmPayload = payload.payload;
    const payer = exactEvmPayload.authorization.from;
    let eip6492Deployment;
    if (payloadV1.scheme !== "exact" || requirements.scheme !== "exact") {
      return {
        isValid: false,
        invalidReason: ErrInvalidScheme,
        payer
      };
    }
    let chainId;
    try {
      chainId = getEvmChainIdV1(payloadV1.network);
    } catch {
      return {
        isValid: false,
        invalidReason: ErrNetworkMismatch,
        payer
      };
    }
    if (!requirements.extra?.name || !requirements.extra?.version) {
      return {
        isValid: false,
        invalidReason: ErrMissingEip712Domain,
        payer
      };
    }
    const { name, version } = requirements.extra;
    const erc20Address = (0, import_viem13.getAddress)(requirements.asset);
    if (payloadV1.network !== requirements.network) {
      return {
        isValid: false,
        invalidReason: ErrNetworkMismatch,
        payer
      };
    }
    const permitTypedData = {
      types: authorizationTypes,
      primaryType: "TransferWithAuthorization",
      domain: {
        name,
        version,
        chainId,
        verifyingContract: erc20Address
      },
      message: {
        from: exactEvmPayload.authorization.from,
        to: exactEvmPayload.authorization.to,
        value: BigInt(exactEvmPayload.authorization.value),
        validAfter: BigInt(exactEvmPayload.authorization.validAfter),
        validBefore: BigInt(exactEvmPayload.authorization.validBefore),
        nonce: exactEvmPayload.authorization.nonce
      }
    };
    let isValid = false;
    try {
      isValid = await this.signer.verifyTypedData({
        address: payer,
        ...permitTypedData,
        signature: exactEvmPayload.signature
      });
    } catch {
      isValid = false;
    }
    const signature = exactEvmPayload.signature;
    const sigLen = signature.startsWith("0x") ? signature.length - 2 : signature.length;
    const erc6492Data = (0, import_viem13.parseErc6492Signature)(signature);
    const hasDeploymentInfo = erc6492Data.address && erc6492Data.data && !(0, import_viem13.isAddressEqual)(erc6492Data.address, "0x0000000000000000000000000000000000000000");
    if (hasDeploymentInfo) {
      eip6492Deployment = {
        factoryAddress: erc6492Data.address,
        factoryCalldata: erc6492Data.data
      };
    }
    if (!isValid) {
      const isSmartWallet = sigLen > 130;
      if (!isSmartWallet) {
        return {
          isValid: false,
          invalidReason: ErrInvalidSignature,
          payer
        };
      }
      const bytecode = await this.signer.getCode({ address: payer });
      const isDeployed = bytecode && bytecode !== "0x";
      if (!isDeployed && !hasDeploymentInfo) {
        return {
          isValid: false,
          invalidReason: ErrUndeployedSmartWallet,
          payer
        };
      }
    }
    if ((0, import_viem13.getAddress)(exactEvmPayload.authorization.to) !== (0, import_viem13.getAddress)(requirements.payTo)) {
      return {
        isValid: false,
        invalidReason: ErrRecipientMismatch,
        payer
      };
    }
    const now = Math.floor(Date.now() / 1e3);
    if (BigInt(exactEvmPayload.authorization.validBefore) < BigInt(now + 6)) {
      return {
        isValid: false,
        invalidReason: ErrValidBeforeExpired,
        payer
      };
    }
    if (BigInt(exactEvmPayload.authorization.validAfter) > BigInt(now)) {
      return {
        isValid: false,
        invalidReason: ErrValidAfterInFuture,
        payer
      };
    }
    if (BigInt(exactEvmPayload.authorization.value) !== BigInt(requirementsV1.maxAmountRequired)) {
      return {
        isValid: false,
        invalidReason: ErrInvalidAuthorizationValue,
        payer
      };
    }
    if (options?.simulate !== false) {
      const simulationSucceeded = await simulateEip3009Transfer(
        this.signer,
        erc20Address,
        exactEvmPayload,
        eip6492Deployment
      );
      if (!simulationSucceeded) {
        return diagnoseEip3009SimulationFailure(
          this.signer,
          erc20Address,
          exactEvmPayload,
          requirements,
          requirementsV1.maxAmountRequired
        );
      }
    }
    return {
      isValid: true,
      invalidReason: void 0,
      payer
    };
  }
};

// src/exact/facilitator/register.ts
function registerExactEvmScheme(facilitator, config) {
  facilitator.register(
    config.networks,
    new ExactEvmScheme(config.signer, {
      eip6492AllowedFactories: config.eip6492AllowedFactories,
      simulateInSettle: config.simulateInSettle
    })
  );
  facilitator.registerV1(
    NETWORKS,
    new ExactEvmSchemeV12(config.signer, {
      eip6492AllowedFactories: config.eip6492AllowedFactories,
      simulateInSettle: config.simulateInSettle
    })
  );
  return facilitator;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExactEvmScheme,
  registerExactEvmScheme
});
