// @x402/core/dist/cjs/schemas/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/schemas/index.ts
var schemas_exports = {};
__export(schemas_exports, {
  Any: () => Any,
  NetworkSchema: () => NetworkSchema,
  NetworkSchemaV1: () => NetworkSchemaV1,
  NetworkSchemaV2: () => NetworkSchemaV2,
  NonEmptyString: () => NonEmptyString,
  OptionalAny: () => OptionalAny,
  PaymentPayloadSchema: () => PaymentPayloadSchema,
  PaymentPayloadV1Schema: () => PaymentPayloadV1Schema,
  PaymentPayloadV2Schema: () => PaymentPayloadV2Schema,
  PaymentRequiredSchema: () => PaymentRequiredSchema,
  PaymentRequiredV1Schema: () => PaymentRequiredV1Schema,
  PaymentRequiredV2Schema: () => PaymentRequiredV2Schema,
  PaymentRequirementsSchema: () => PaymentRequirementsSchema,
  PaymentRequirementsV1Schema: () => PaymentRequirementsV1Schema,
  PaymentRequirementsV2Schema: () => PaymentRequirementsV2Schema,
  ResourceInfoSchema: () => ResourceInfoSchema,
  isPaymentPayload: () => isPaymentPayload,
  isPaymentPayloadV1: () => isPaymentPayloadV1,
  isPaymentPayloadV2: () => isPaymentPayloadV2,
  isPaymentRequired: () => isPaymentRequired,
  isPaymentRequiredV1: () => isPaymentRequiredV1,
  isPaymentRequiredV2: () => isPaymentRequiredV2,
  isPaymentRequirements: () => isPaymentRequirements,
  isPaymentRequirementsV1: () => isPaymentRequirementsV1,
  isPaymentRequirementsV2: () => isPaymentRequirementsV2,
  parsePaymentPayload: () => parsePaymentPayload,
  parsePaymentRequired: () => parsePaymentRequired,
  parsePaymentRequirements: () => parsePaymentRequirements,
  validatePaymentPayload: () => validatePaymentPayload,
  validatePaymentRequired: () => validatePaymentRequired,
  validatePaymentRequirements: () => validatePaymentRequirements,
  z: () => import_zod2.z
});
module.exports = __toCommonJS(schemas_exports);
var import_zod = require("zod");
var import_zod2 = require("zod");
var NonEmptyString = import_zod.z.string().min(1);
var Any = import_zod.z.record(import_zod.z.unknown());
var OptionalAny = import_zod.z.record(import_zod.z.unknown()).optional().nullable();
var NetworkSchemaV1 = NonEmptyString;
var NetworkSchemaV2 = import_zod.z.string().min(3).refine((val) => val.includes(":"), {
  message: "Network must be in CAIP-2 format (e.g., 'eip155:84532')"
});
var NetworkSchema = import_zod.z.union([NetworkSchemaV1, NetworkSchemaV2]);
var PRINTABLE_ASCII_REGEX = /^[\x20-\x7e]+$/;
var ResourceInfoSchema = import_zod.z.object({
  url: NonEmptyString,
  description: import_zod.z.string().optional(),
  mimeType: import_zod.z.string().optional(),
  serviceName: import_zod.z.string().min(1).max(32).regex(PRINTABLE_ASCII_REGEX).optional(),
  tags: import_zod.z.array(import_zod.z.string().min(1).max(32).regex(PRINTABLE_ASCII_REGEX)).max(5).optional(),
  iconUrl: import_zod.z.string().max(2048).optional()
});
var PaymentRequirementsV1Schema = import_zod.z.object({
  scheme: NonEmptyString,
  network: NetworkSchemaV1,
  maxAmountRequired: NonEmptyString,
  resource: NonEmptyString,
  // URL string in V1
  description: import_zod.z.string(),
  mimeType: import_zod.z.string().optional(),
  outputSchema: Any.optional().nullable(),
  payTo: NonEmptyString,
  maxTimeoutSeconds: import_zod.z.number().positive(),
  asset: NonEmptyString,
  extra: OptionalAny
});
var PaymentRequiredV1Schema = import_zod.z.object({
  x402Version: import_zod.z.literal(1),
  error: import_zod.z.string().optional(),
  accepts: import_zod.z.array(PaymentRequirementsV1Schema).min(1)
});
var PaymentPayloadV1Schema = import_zod.z.object({
  x402Version: import_zod.z.literal(1),
  scheme: NonEmptyString,
  network: NetworkSchemaV1,
  payload: Any
});
var PaymentRequirementsV2Schema = import_zod.z.object({
  scheme: NonEmptyString,
  network: NetworkSchemaV2,
  amount: NonEmptyString,
  asset: NonEmptyString,
  payTo: NonEmptyString,
  maxTimeoutSeconds: import_zod.z.number().positive(),
  extra: OptionalAny
});
var PaymentRequiredV2Schema = import_zod.z.object({
  x402Version: import_zod.z.literal(2),
  error: import_zod.z.string().optional(),
  resource: ResourceInfoSchema,
  accepts: import_zod.z.array(PaymentRequirementsV2Schema).min(1),
  extensions: OptionalAny
});
var PaymentPayloadV2Schema = import_zod.z.object({
  x402Version: import_zod.z.literal(2),
  resource: ResourceInfoSchema.optional(),
  accepted: PaymentRequirementsV2Schema,
  payload: Any,
  extensions: OptionalAny
});
var PaymentRequirementsSchema = import_zod.z.union([
  PaymentRequirementsV1Schema,
  PaymentRequirementsV2Schema
]);
var PaymentRequiredSchema = import_zod.z.discriminatedUnion("x402Version", [
  PaymentRequiredV1Schema,
  PaymentRequiredV2Schema
]);
var PaymentPayloadSchema = import_zod.z.discriminatedUnion("x402Version", [
  PaymentPayloadV1Schema,
  PaymentPayloadV2Schema
]);
function parsePaymentRequired(value) {
  return PaymentRequiredSchema.safeParse(value);
}
function validatePaymentRequired(value) {
  return PaymentRequiredSchema.parse(value);
}
function isPaymentRequired(value) {
  return PaymentRequiredSchema.safeParse(value).success;
}
function parsePaymentRequirements(value) {
  return PaymentRequirementsSchema.safeParse(value);
}
function validatePaymentRequirements(value) {
  return PaymentRequirementsSchema.parse(value);
}
function isPaymentRequirements(value) {
  return PaymentRequirementsSchema.safeParse(value).success;
}
function parsePaymentPayload(value) {
  return PaymentPayloadSchema.safeParse(value);
}
function validatePaymentPayload(value) {
  return PaymentPayloadSchema.parse(value);
}
function isPaymentPayload(value) {
  return PaymentPayloadSchema.safeParse(value).success;
}
function isPaymentRequiredV1(value) {
  return PaymentRequiredV1Schema.safeParse(value).success;
}
function isPaymentRequiredV2(value) {
  return PaymentRequiredV2Schema.safeParse(value).success;
}
function isPaymentRequirementsV1(value) {
  return PaymentRequirementsV1Schema.safeParse(value).success;
}
function isPaymentRequirementsV2(value) {
  return PaymentRequirementsV2Schema.safeParse(value).success;
}
function isPaymentPayloadV1(value) {
  return PaymentPayloadV1Schema.safeParse(value).success;
}
function isPaymentPayloadV2(value) {
  return PaymentPayloadV2Schema.safeParse(value).success;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Any,
  NetworkSchema,
  NetworkSchemaV1,
  NetworkSchemaV2,
  NonEmptyString,
  OptionalAny,
  PaymentPayloadSchema,
  PaymentPayloadV1Schema,
  PaymentPayloadV2Schema,
  PaymentRequiredSchema,
  PaymentRequiredV1Schema,
  PaymentRequiredV2Schema,
  PaymentRequirementsSchema,
  PaymentRequirementsV1Schema,
  PaymentRequirementsV2Schema,
  ResourceInfoSchema,
  isPaymentPayload,
  isPaymentPayloadV1,
  isPaymentPayloadV2,
  isPaymentRequired,
  isPaymentRequiredV1,
  isPaymentRequiredV2,
  isPaymentRequirements,
  isPaymentRequirementsV1,
  isPaymentRequirementsV2,
  parsePaymentPayload,
  parsePaymentRequired,
  parsePaymentRequirements,
  validatePaymentPayload,
  validatePaymentRequired,
  validatePaymentRequirements,
  z
});
