// @x402/extensions/dist/esm/sign-in-with-x/index.mjs
import {
  InMemorySIWxStorage,
  SIGN_IN_WITH_X,
  SIWxPayloadSchema,
  SOLANA_DEVNET,
  SOLANA_MAINNET,
  SOLANA_TESTNET,
  buildSIWxSchema,
  createSIWxClientExtension,
  createSIWxClientHook,
  createSIWxMessage,
  createSIWxPayload,
  createSIWxRequestHook,
  createSIWxResourceServerExtension,
  createSIWxSettleHook,
  declareSIWxExtension,
  decodeBase58,
  encodeBase58,
  encodeSIWxHeader,
  extractEVMChainId,
  extractSolanaChainReference,
  formatSIWEMessage,
  formatSIWSMessage,
  getEVMAddress,
  getSolanaAddress,
  isEVMSigner,
  isSolanaSigner,
  parseSIWxHeader,
  signEVMMessage,
  signSolanaMessage,
  validateSIWxMessage,
  verifyEVMSignature,
  verifySIWxSignature,
  verifySolanaSignature,
  wrapFetchWithSIWx
} from "../chunk-LMLJI6VE.mjs";
export {
  InMemorySIWxStorage,
  SIGN_IN_WITH_X,
  SIWxPayloadSchema,
  SOLANA_DEVNET,
  SOLANA_MAINNET,
  SOLANA_TESTNET,
  buildSIWxSchema,
  createSIWxClientExtension,
  createSIWxClientHook,
  createSIWxMessage,
  createSIWxPayload,
  createSIWxRequestHook,
  createSIWxResourceServerExtension,
  createSIWxSettleHook,
  declareSIWxExtension,
  decodeBase58,
  encodeBase58,
  encodeSIWxHeader,
  extractEVMChainId,
  extractSolanaChainReference,
  formatSIWEMessage,
  formatSIWSMessage,
  getEVMAddress,
  getSolanaAddress,
  isEVMSigner,
  isSolanaSigner,
  parseSIWxHeader,
  signEVMMessage,
  signSolanaMessage,
  validateSIWxMessage,
  verifyEVMSignature,
  verifySIWxSignature,
  verifySolanaSignature,
  wrapFetchWithSIWx
};
