// @x402/evm/dist/cjs/exact/server/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/exact/server/index.ts
var server_exports = {};
__export(server_exports, {
  ExactEvmScheme: () => ExactEvmScheme,
  registerExactEvmScheme: () => registerExactEvmScheme
});
module.exports = __toCommonJS(server_exports);

// src/exact/server/scheme.ts
var import_utils = require("@x402/core/utils");

// src/shared/defaultAssets.ts
var DEFAULT_STABLECOINS = {
  "eip155:8453": {
    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    name: "USD Coin",
    version: "2",
    decimals: 6
  },
  // Base mainnet USDC
  "eip155:84532": {
    address: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
    name: "USDC",
    version: "2",
    decimals: 6
  },
  // Base Sepolia USDC
  "eip155:4326": {
    address: "0xFAfDdbb3FC7688494971a79cc65DCa3EF82079E7",
    name: "MegaUSD",
    version: "1",
    decimals: 18,
    assetTransferMethod: "permit2",
    supportsEip2612: true
  },
  // MegaETH mainnet MegaUSD (no EIP-3009, supports EIP-2612)
  "eip155:143": {
    address: "0x754704Bc059F8C67012fEd69BC8A327a5aafb603",
    name: "USD Coin",
    version: "2",
    decimals: 6
  },
  // Monad mainnet USDC
  "eip155:988": {
    address: "0x779Ded0c9e1022225f8E0630b35a9b54bE713736",
    name: "USDT0",
    version: "1",
    decimals: 6
  },
  // Stable mainnet USDT0
  "eip155:2201": {
    address: "0x78Cf24370174180738C5B8E352B6D14c83a6c9A9",
    name: "USDT0",
    version: "1",
    decimals: 6
  },
  // Stable testnet USDT0
  "eip155:137": {
    address: "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359",
    name: "USD Coin",
    version: "2",
    decimals: 6
  },
  // Polygon mainnet USDC
  "eip155:42161": {
    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
    name: "USD Coin",
    version: "2",
    decimals: 6
  },
  // Arbitrum One USDC
  "eip155:421614": {
    address: "0x75faf114eafb1BDbe2F0316DF893fd58CE46AA4d",
    name: "USD Coin",
    version: "2",
    decimals: 6
  },
  // Arbitrum Sepolia USDC
  "eip155:31612": {
    address: "0xdD468A1DDc392dcdbEf6db6e34E89AA338F9F186",
    name: "Mezo USD",
    version: "1",
    decimals: 18,
    assetTransferMethod: "permit2",
    supportsEip2612: true
  },
  // Mezo mainnet mUSD (no EIP-3009, supports EIP-2612)
  "eip155:31611": {
    address: "0x118917a40FAF1CD7a13dB0Ef56C86De7973Ac503",
    name: "Mezo USD",
    version: "1",
    decimals: 18,
    assetTransferMethod: "permit2",
    supportsEip2612: true
  },
  // Mezo Testnet mUSD (no EIP-3009, supports EIP-2612)
  "eip155:723487": {
    address: "0x33ad9e4BD16B69B5BFdED37D8B5D9fF9aba014Fb",
    name: "Stable Coin",
    version: "1",
    decimals: 6,
    assetTransferMethod: "permit2",
    supportsEip2612: true
  },
  // Radius Network SBC (no EIP-3009, supports EIP-2612)
  "eip155:72344": {
    address: "0x33ad9e4BD16B69B5BFdED37D8B5D9fF9aba014Fb",
    name: "Stable Coin",
    version: "1",
    decimals: 6,
    assetTransferMethod: "permit2",
    supportsEip2612: true
  },
  // Radius Testnet SBC (no EIP-3009, supports EIP-2612)
  "eip155:36900": {
    address: "0x9cb8142aEBBcdc60AF7c97Af897A67A8f3CA71C2",
    name: "USDC.e",
    version: "2",
    decimals: 6
  },
  // ADI Chain USDC.e (EIP-3009 supported)
  "eip155:190415": {
    address: "0x401eCb1D350407f13ba348573E5630B83638E30D",
    name: "Bridged USDC",
    version: "2",
    decimals: 6
  },
  // HPP mainnet USDC.e
  "eip155:181228": {
    address: "0x401eCb1D350407f13ba348573E5630B83638E30D",
    name: "Bridged USDC",
    version: "2",
    decimals: 6
  },
  // HPP Sepolia USDC.e
  "eip155:50": {
    address: "0xfA2958CB79b0491CC627c1557F441eF849Ca8eb1",
    name: "USDC",
    version: "2",
    decimals: 6
  },
  // XDC Network mainnet USDC (Bridged USDC Standard, EIP-3009 supported)
  "eip155:51": {
    address: "0xb5AB69F7bBada22B28e79C8FFAECe55eF1c771D4",
    name: "USDC",
    version: "2",
    decimals: 6
  }
  // XDC Apothem testnet USDC (Bridged USDC Standard, EIP-3009 supported)
};
function getDefaultAsset(network) {
  const info = DEFAULT_STABLECOINS[network];
  if (!info) {
    throw new Error(`No default asset configured for network ${network}`);
  }
  return info;
}

// src/exact/server/scheme.ts
var ExactEvmScheme = class {
  constructor() {
    this.scheme = "exact";
    this.moneyParsers = [];
  }
  /**
   * Register a custom money parser in the parser chain.
   * Multiple parsers can be registered - they will be tried in registration order.
   * Each parser receives a decimal amount (e.g., 1.50 for $1.50).
   * If a parser returns null, the next parser in the chain will be tried.
   * The default parser is always the final fallback.
   *
   * @param parser - Custom function to convert amount to AssetAmount (or null to skip)
   * @returns The server instance for chaining
   *
   * @example
   * evmServer.registerMoneyParser(async (amount, network) => {
   *   // Custom conversion logic
   *   if (amount > 100) {
   *     // Use different token for large amounts
   *     return { amount: (amount * 1e18).toString(), asset: "0xCustomToken" };
   *   }
   *   return null; // Use next parser
   * });
   */
  registerMoneyParser(parser) {
    this.moneyParsers.push(parser);
    return this;
  }
  /**
   * Returns the decimal precision of the default stablecoin for the given network.
   * Implements the optional AssetDecimalsProvider interface used by resolveSettlementOverrideAmount.
   *
   * @param _asset - The asset symbol (unused; defaults to the network's default stablecoin)
   * @param network - The network to look up the default asset for
   * @returns The number of decimal places for the asset
   */
  getAssetDecimals(_asset, network) {
    try {
      return getDefaultAsset(network).decimals;
    } catch {
      return 6;
    }
  }
  /**
   * Parses a price into an asset amount.
   * If price is already an AssetAmount, returns it directly.
   * If price is Money (string | number), parses to decimal and tries custom parsers.
   * Falls back to default conversion if all custom parsers return null.
   *
   * @param price - The price to parse
   * @param network - The network to use
   * @returns Promise that resolves to the parsed asset amount
   */
  async parsePrice(price, network) {
    if (typeof price === "object" && price !== null && "amount" in price) {
      if (!price.asset) {
        throw new Error(`Asset address must be specified for AssetAmount on network ${network}`);
      }
      return {
        amount: price.amount,
        asset: price.asset,
        extra: price.extra || {}
      };
    }
    const amount = this.parseMoneyToDecimal(price);
    for (const parser of this.moneyParsers) {
      const result = await parser(amount, network);
      if (result !== null) {
        return result;
      }
    }
    return this.defaultMoneyConversion(amount, network);
  }
  /**
   * Build payment requirements for this scheme/network combination
   *
   * @param paymentRequirements - The base payment requirements
   * @param supportedKind - The supported kind from facilitator (unused)
   * @param supportedKind.x402Version - The x402 version
   * @param supportedKind.scheme - The logical payment scheme
   * @param supportedKind.network - The network identifier in CAIP-2 format
   * @param supportedKind.extra - Optional extra metadata regarding scheme/network implementation details
   * @param extensionKeys - Extension keys supported by the facilitator (unused)
   * @returns Payment requirements ready to be sent to clients
   */
  enhancePaymentRequirements(paymentRequirements, supportedKind, extensionKeys) {
    void supportedKind;
    void extensionKeys;
    return Promise.resolve(paymentRequirements);
  }
  /**
   * Parse Money (string | number) to a decimal number.
   * Handles formats like "$1.50", "1.50", 1.50, etc.
   *
   * @param money - The money value to parse
   * @returns Decimal number
   */
  parseMoneyToDecimal(money) {
    if (typeof money === "number") {
      return money;
    }
    return (0, import_utils.parseMoneyString)(money);
  }
  /**
   * Converts a numeric dollar amount to an AssetAmount using the default token for the network.
   *
   * @param amount - The dollar amount as a number
   * @param network - The target network
   * @returns The converted asset amount with token metadata
   */
  defaultMoneyConversion(amount, network) {
    const assetInfo = getDefaultAsset(network);
    const tokenAmount = (0, import_utils.convertToTokenAmount)((0, import_utils.numberToDecimalString)(amount), assetInfo.decimals);
    const includeEip712Domain = !assetInfo.assetTransferMethod || assetInfo.supportsEip2612;
    return {
      amount: tokenAmount,
      asset: assetInfo.address,
      extra: {
        ...includeEip712Domain && {
          name: assetInfo.name,
          version: assetInfo.version
        },
        ...assetInfo.assetTransferMethod && {
          assetTransferMethod: assetInfo.assetTransferMethod
        }
      }
    };
  }
};

// src/exact/server/register.ts
function registerExactEvmScheme(server, config = {}) {
  if (config.networks && config.networks.length > 0) {
    config.networks.forEach((network) => {
      server.register(network, new ExactEvmScheme());
    });
  } else {
    server.register("eip155:*", new ExactEvmScheme());
  }
  return server;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ExactEvmScheme,
  registerExactEvmScheme
});
