// @x402/core/dist/cjs/utils/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/index.ts
var utils_exports = {};
__export(utils_exports, {
  Base64EncodedRegex: () => Base64EncodedRegex,
  convertToTokenAmount: () => convertToTokenAmount,
  deepEqual: () => deepEqual,
  findByNetworkAndScheme: () => findByNetworkAndScheme,
  findFacilitatorBySchemeAndNetwork: () => findFacilitatorBySchemeAndNetwork,
  findSchemesByNetwork: () => findSchemesByNetwork,
  networkMatchesPattern: () => networkMatchesPattern,
  numberToDecimalString: () => numberToDecimalString,
  parseMoneyString: () => parseMoneyString,
  safeBase64Decode: () => safeBase64Decode,
  safeBase64Encode: () => safeBase64Encode
});
module.exports = __toCommonJS(utils_exports);
function numberToDecimalString(n) {
  const str = n.toString();
  if (!/[eE]/.test(str)) return str;
  const [significand, exponentStr] = str.split(/[eE]/);
  const exp = parseInt(exponentStr, 10);
  const negative = significand.startsWith("-");
  const abs = negative ? significand.slice(1) : significand;
  const [intDigits, fracDigits = ""] = abs.split(".");
  const allDigits = intDigits + fracDigits;
  const decimalPos = intDigits.length + exp;
  let result;
  if (decimalPos <= 0) {
    result = "0." + "0".repeat(-decimalPos) + allDigits;
  } else if (decimalPos >= allDigits.length) {
    result = allDigits + "0".repeat(decimalPos - allDigits.length);
  } else {
    result = allDigits.slice(0, decimalPos) + "." + allDigits.slice(decimalPos);
  }
  return (negative ? "-" : "") + result;
}
function parseMoneyString(money) {
  const cleaned = money.replace(/^\$/, "").trim();
  if (!/^-?\d+(?:\.\d+)?$/.test(cleaned) || /[eE]/.test(cleaned)) {
    throw new Error(`Invalid money format: ${money}`);
  }
  const amount = Number(cleaned);
  if (!Number.isFinite(amount) || amount < 0) {
    throw new Error(`Invalid money format: ${money}`);
  }
  return amount;
}
function convertToTokenAmount(decimalAmount, decimals) {
  if (/[eE]/.test(decimalAmount)) {
    throw new Error(
      `Invalid amount: ${decimalAmount} \u2014 use decimal notation, not scientific notation`
    );
  }
  if (!/^-?\d+\.?\d*$/.test(decimalAmount)) {
    throw new Error(`Invalid amount: ${decimalAmount}`);
  }
  const [intPart, decPart = ""] = decimalAmount.split(".");
  const paddedDec = decPart.padEnd(decimals, "0").slice(0, decimals);
  const tokenAmount = (intPart + paddedDec).replace(/^0+/, "") || "0";
  if (tokenAmount === "0" && /[1-9]/.test(decimalAmount)) {
    throw new Error(
      `Amount ${decimalAmount} is too small to represent with ${decimals} decimal places`
    );
  }
  return tokenAmount;
}
var escapeRegExp = (value) => value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
var networkPatternToRegExp = (pattern) => {
  const source = escapeRegExp(pattern).replace(/\\\*/g, ".*");
  return new RegExp(`^${source}$`);
};
var networkMatchesPattern = (pattern, network) => {
  return networkPatternToRegExp(pattern).test(network);
};
var findSchemesByNetwork = (map, network) => {
  let implementationsByScheme = map.get(network);
  if (!implementationsByScheme) {
    for (const [registeredNetworkPattern, implementations] of map.entries()) {
      if (networkMatchesPattern(registeredNetworkPattern, network)) {
        implementationsByScheme = implementations;
        break;
      }
    }
  }
  return implementationsByScheme;
};
var findByNetworkAndScheme = (map, scheme, network) => {
  return findSchemesByNetwork(map, network)?.get(scheme);
};
var findFacilitatorBySchemeAndNetwork = (schemeMap, scheme, network) => {
  const schemeData = schemeMap.get(scheme);
  if (!schemeData) return void 0;
  if (schemeData.networks.has(network)) {
    return schemeData.facilitator;
  }
  if (networkMatchesPattern(schemeData.pattern, network)) {
    return schemeData.facilitator;
  }
  return void 0;
};
var Base64EncodedRegex = /^[A-Za-z0-9+/]*={0,2}$/;
function safeBase64Encode(data) {
  if (typeof globalThis !== "undefined" && typeof globalThis.btoa === "function") {
    const bytes = new TextEncoder().encode(data);
    const binaryString = Array.from(bytes, (byte) => String.fromCharCode(byte)).join("");
    return globalThis.btoa(binaryString);
  }
  return Buffer.from(data, "utf8").toString("base64");
}
function safeBase64Decode(data) {
  if (typeof globalThis !== "undefined" && typeof globalThis.atob === "function") {
    const binaryString = globalThis.atob(data);
    const bytes = new Uint8Array(binaryString.length);
    for (let i = 0; i < binaryString.length; i++) {
      bytes[i] = binaryString.charCodeAt(i);
    }
    const decoder = new TextDecoder("utf-8");
    return decoder.decode(bytes);
  }
  return Buffer.from(data, "base64").toString("utf-8");
}
function deepEqual(obj1, obj2) {
  const normalize = (obj) => {
    if (obj === null || obj === void 0) return JSON.stringify(obj);
    if (typeof obj !== "object") return JSON.stringify(obj);
    if (Array.isArray(obj)) {
      return JSON.stringify(
        obj.map(
          (item) => typeof item === "object" && item !== null ? JSON.parse(normalize(item)) : item
        )
      );
    }
    const sorted = {};
    Object.keys(obj).sort().forEach((key) => {
      const value = obj[key];
      sorted[key] = typeof value === "object" && value !== null ? JSON.parse(normalize(value)) : value;
    });
    return JSON.stringify(sorted);
  };
  try {
    return normalize(obj1) === normalize(obj2);
  } catch {
    return JSON.stringify(obj1) === JSON.stringify(obj2);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Base64EncodedRegex,
  convertToTokenAmount,
  deepEqual,
  findByNetworkAndScheme,
  findFacilitatorBySchemeAndNetwork,
  findSchemesByNetwork,
  networkMatchesPattern,
  numberToDecimalString,
  parseMoneyString,
  safeBase64Decode,
  safeBase64Encode
});
