// @x402/evm/dist/esm/upto/facilitator/index.mjs
import {
  isUptoPermit2Payload
} from "../../chunk-JK7SLLF7.mjs";
import "../../chunk-U4HCGTLU.mjs";
import {
  ErrUptoFacilitatorMismatch,
  ErrUptoInvalidScheme,
  ErrUptoNetworkMismatch,
  ErrUptoSettlementExceedsAmount,
  buildUptoPermit2SettleArgs,
  checkPermit2Prerequisites,
  diagnosePermit2SimulationFailure,
  mapSettleError,
  simulatePermit2Settle,
  simulatePermit2SettleWithErc20Approval,
  simulatePermit2SettleWithPermit,
  splitEip2612Signature,
  validateEip2612PermitForPayment,
  validateErc20ApprovalForPayment,
  waitAndReturnSettleResponse
} from "../../chunk-CE5P6FKO.mjs";
import {
  ErrAssetNotDeployedContract,
  ErrPermit2AmountMismatch
} from "../../chunk-P3QOX3QZ.mjs";
import {
  ERC20_APPROVAL_GAS_SPONSORING_KEY,
  appendDataSuffix,
  extractEip2612GasSponsoringInfo,
  extractErc20ApprovalGasSponsoringInfo,
  resolveDataSuffix,
  resolveErc20ApprovalExtensionSigner
} from "../../chunk-27MWX225.mjs";
import {
  PERMIT2_ADDRESS,
  uptoPermit2WitnessTypes,
  x402UptoPermit2ProxyABI,
  x402UptoPermit2ProxyAddress
} from "../../chunk-MACPBXCT.mjs";
import "../../chunk-VS3RYAYE.mjs";
import {
  getEvmChainId
} from "../../chunk-TW7Z65AO.mjs";

// src/upto/facilitator/permit2.ts
import { getAddress, encodeFunctionData } from "viem";
var uptoProxyConfig = {
  proxyAddress: x402UptoPermit2ProxyAddress,
  proxyABI: x402UptoPermit2ProxyABI
};
async function verifyUptoPermit2(signer, payload, requirements, permit2Payload, context, options) {
  const payer = permit2Payload.permit2Authorization.from;
  if (payload.accepted.scheme !== "upto" || requirements.scheme !== "upto") {
    return {
      isValid: false,
      invalidReason: ErrUptoInvalidScheme,
      payer
    };
  }
  if (payload.accepted.network !== requirements.network) {
    return {
      isValid: false,
      invalidReason: ErrUptoNetworkMismatch,
      payer
    };
  }
  const chainId = getEvmChainId(requirements.network);
  const tokenAddress = getAddress(requirements.asset);
  const assetBytecode = await signer.getCode({ address: tokenAddress });
  if (!assetBytecode || assetBytecode === "0x") {
    return { isValid: false, invalidReason: ErrAssetNotDeployedContract, payer };
  }
  if (getAddress(permit2Payload.permit2Authorization.spender) !== getAddress(x402UptoPermit2ProxyAddress)) {
    return {
      isValid: false,
      invalidReason: "invalid_permit2_spender",
      payer
    };
  }
  if (getAddress(permit2Payload.permit2Authorization.witness.to) !== getAddress(requirements.payTo)) {
    return {
      isValid: false,
      invalidReason: "invalid_permit2_recipient_mismatch",
      payer
    };
  }
  const facilitatorAddresses = signer.getAddresses();
  const witnessFacilitator = getAddress(permit2Payload.permit2Authorization.witness.facilitator);
  const isFacilitatorMatch = facilitatorAddresses.some(
    (addr) => getAddress(addr) === witnessFacilitator
  );
  if (!isFacilitatorMatch) {
    return {
      isValid: false,
      invalidReason: ErrUptoFacilitatorMismatch,
      payer
    };
  }
  const now = Math.floor(Date.now() / 1e3);
  if (BigInt(permit2Payload.permit2Authorization.deadline) < BigInt(now + 6)) {
    return {
      isValid: false,
      invalidReason: "permit2_deadline_expired",
      payer
    };
  }
  if (BigInt(permit2Payload.permit2Authorization.witness.validAfter) > BigInt(now)) {
    return {
      isValid: false,
      invalidReason: "permit2_not_yet_valid",
      payer
    };
  }
  if (BigInt(permit2Payload.permit2Authorization.permitted.amount) !== BigInt(requirements.amount)) {
    return {
      isValid: false,
      invalidReason: ErrPermit2AmountMismatch,
      payer
    };
  }
  if (getAddress(permit2Payload.permit2Authorization.permitted.token) !== tokenAddress) {
    return {
      isValid: false,
      invalidReason: "permit2_token_mismatch",
      payer
    };
  }
  const permit2TypedData = {
    types: uptoPermit2WitnessTypes,
    primaryType: "PermitWitnessTransferFrom",
    domain: {
      name: "Permit2",
      chainId,
      verifyingContract: PERMIT2_ADDRESS
    },
    message: {
      permitted: {
        token: getAddress(permit2Payload.permit2Authorization.permitted.token),
        amount: BigInt(permit2Payload.permit2Authorization.permitted.amount)
      },
      spender: getAddress(permit2Payload.permit2Authorization.spender),
      nonce: BigInt(permit2Payload.permit2Authorization.nonce),
      deadline: BigInt(permit2Payload.permit2Authorization.deadline),
      witness: {
        to: getAddress(permit2Payload.permit2Authorization.witness.to),
        facilitator: getAddress(permit2Payload.permit2Authorization.witness.facilitator),
        validAfter: BigInt(permit2Payload.permit2Authorization.witness.validAfter)
      }
    }
  };
  let signatureValid = false;
  try {
    signatureValid = await signer.verifyTypedData({
      address: payer,
      ...permit2TypedData,
      signature: permit2Payload.signature
    });
  } catch {
    signatureValid = false;
  }
  if (!signatureValid) {
    const bytecode = await signer.getCode({ address: payer });
    const isDeployedContract = bytecode && bytecode !== "0x";
    if (!isDeployedContract) {
      return {
        isValid: false,
        invalidReason: "invalid_permit2_signature",
        payer
      };
    }
  }
  if (options?.simulate === false) {
    return { isValid: true, invalidReason: void 0, payer };
  }
  const facilitatorAddress = getAddress(permit2Payload.permit2Authorization.witness.facilitator);
  const uptoSettleArgs = buildUptoPermit2SettleArgs(
    permit2Payload,
    BigInt(requirements.amount),
    facilitatorAddress
  );
  const eip2612InfoForSim = extractEip2612GasSponsoringInfo(payload);
  if (eip2612InfoForSim) {
    const fieldResult = validateEip2612PermitForPayment(eip2612InfoForSim, payer, tokenAddress);
    if (!fieldResult.isValid) {
      return { isValid: false, invalidReason: fieldResult.invalidReason, payer };
    }
    const simOk2 = await simulatePermit2SettleWithPermit(
      uptoProxyConfig,
      signer,
      uptoSettleArgs,
      eip2612InfoForSim
    );
    if (!simOk2) {
      return diagnosePermit2SimulationFailure(
        uptoProxyConfig,
        signer,
        tokenAddress,
        permit2Payload,
        requirements.amount
      );
    }
    return { isValid: true, invalidReason: void 0, payer };
  }
  const erc20GasSponsorshipExtension = context?.getExtension(
    ERC20_APPROVAL_GAS_SPONSORING_KEY
  );
  if (erc20GasSponsorshipExtension) {
    const erc20Info = extractErc20ApprovalGasSponsoringInfo(payload);
    if (erc20Info) {
      const fieldResult = await validateErc20ApprovalForPayment(erc20Info, payer, tokenAddress);
      if (!fieldResult.isValid) {
        return { isValid: false, invalidReason: fieldResult.invalidReason, payer };
      }
      const extensionSigner = resolveErc20ApprovalExtensionSigner(
        erc20GasSponsorshipExtension,
        requirements.network
      );
      if (extensionSigner?.simulateTransactions) {
        const simOk2 = await simulatePermit2SettleWithErc20Approval(
          uptoProxyConfig,
          extensionSigner,
          uptoSettleArgs,
          erc20Info
        );
        if (!simOk2) {
          return diagnosePermit2SimulationFailure(
            uptoProxyConfig,
            signer,
            tokenAddress,
            permit2Payload,
            requirements.amount
          );
        }
        return { isValid: true, invalidReason: void 0, payer };
      }
      return checkPermit2Prerequisites(
        uptoProxyConfig,
        signer,
        tokenAddress,
        payer,
        requirements.amount
      );
    }
  }
  const simOk = await simulatePermit2Settle(uptoProxyConfig, signer, uptoSettleArgs);
  if (!simOk) {
    return diagnosePermit2SimulationFailure(
      uptoProxyConfig,
      signer,
      tokenAddress,
      permit2Payload,
      requirements.amount
    );
  }
  return {
    isValid: true,
    invalidReason: void 0,
    payer
  };
}
async function settleUptoPermit2(signer, payload, requirements, permit2Payload, context, config) {
  const payer = permit2Payload.permit2Authorization.from;
  const settlementAmount = BigInt(requirements.amount);
  const verifyRequirements = {
    ...requirements,
    amount: permit2Payload.permit2Authorization.permitted.amount
  };
  const valid = await verifyUptoPermit2(
    signer,
    payload,
    verifyRequirements,
    permit2Payload,
    context,
    { simulate: config?.simulateInSettle ?? true }
  );
  if (!valid.isValid) {
    return {
      success: false,
      network: payload.accepted.network,
      transaction: "",
      errorReason: valid.invalidReason ?? "invalid_scheme",
      payer
    };
  }
  if (settlementAmount === 0n) {
    return {
      success: true,
      transaction: "",
      network: payload.accepted.network,
      payer,
      amount: "0"
    };
  }
  if (settlementAmount > BigInt(permit2Payload.permit2Authorization.permitted.amount)) {
    return {
      success: false,
      network: payload.accepted.network,
      transaction: "",
      errorReason: ErrUptoSettlementExceedsAmount,
      payer
    };
  }
  const facilitatorAddress = getAddress(permit2Payload.permit2Authorization.witness.facilitator);
  const dataSuffix = await resolveDataSuffix(context, {
    paymentPayload: payload,
    paymentRequirements: requirements
  });
  const eip2612Info = extractEip2612GasSponsoringInfo(payload);
  if (eip2612Info) {
    return settleUptoWithEIP2612(
      signer,
      payload,
      permit2Payload,
      eip2612Info,
      settlementAmount,
      facilitatorAddress,
      dataSuffix
    );
  }
  const erc20Info = extractErc20ApprovalGasSponsoringInfo(payload);
  if (erc20Info) {
    const erc20GasSponsorshipExtension = context?.getExtension(
      ERC20_APPROVAL_GAS_SPONSORING_KEY
    );
    const extensionSigner = resolveErc20ApprovalExtensionSigner(
      erc20GasSponsorshipExtension,
      payload.accepted.network
    );
    if (extensionSigner) {
      return settleUptoWithERC20Approval(
        extensionSigner,
        payload,
        permit2Payload,
        erc20Info,
        settlementAmount,
        facilitatorAddress,
        dataSuffix
      );
    }
  }
  return settleUptoDirect(
    signer,
    payload,
    permit2Payload,
    settlementAmount,
    facilitatorAddress,
    dataSuffix
  );
}
async function settleUptoWithEIP2612(signer, payload, permit2Payload, eip2612Info, settlementAmount, facilitatorAddress, dataSuffix) {
  const payer = permit2Payload.permit2Authorization.from;
  try {
    const { v, r, s } = splitEip2612Signature(eip2612Info.signature);
    const tx = await signer.writeContract({
      address: uptoProxyConfig.proxyAddress,
      abi: uptoProxyConfig.proxyABI,
      functionName: "settleWithPermit",
      args: [
        {
          value: BigInt(eip2612Info.amount),
          deadline: BigInt(eip2612Info.deadline),
          r,
          s,
          v
        },
        ...buildUptoPermit2SettleArgs(permit2Payload, settlementAmount, facilitatorAddress)
      ],
      dataSuffix
    });
    const response = await waitAndReturnSettleResponse(signer, tx, payload, payer);
    return { ...response, amount: settlementAmount.toString() };
  } catch (error) {
    return mapSettleError(error, payload, payer);
  }
}
async function settleUptoWithERC20Approval(extensionSigner, payload, permit2Payload, erc20Info, settlementAmount, facilitatorAddress, dataSuffix) {
  const payer = permit2Payload.permit2Authorization.from;
  try {
    const settleData = appendDataSuffix(
      encodeFunctionData({
        abi: uptoProxyConfig.proxyABI,
        functionName: "settle",
        args: buildUptoPermit2SettleArgs(permit2Payload, settlementAmount, facilitatorAddress)
      }),
      dataSuffix
    );
    const txHashes = await extensionSigner.sendTransactions([
      erc20Info.signedTransaction,
      { to: uptoProxyConfig.proxyAddress, data: settleData, gas: BigInt(3e5) }
    ]);
    const settleTxHash = txHashes[txHashes.length - 1];
    const response = await waitAndReturnSettleResponse(
      extensionSigner,
      settleTxHash,
      payload,
      payer
    );
    return { ...response, amount: settlementAmount.toString() };
  } catch (error) {
    return mapSettleError(error, payload, payer);
  }
}
async function settleUptoDirect(signer, payload, permit2Payload, settlementAmount, facilitatorAddress, dataSuffix) {
  const payer = permit2Payload.permit2Authorization.from;
  try {
    const tx = await signer.writeContract({
      address: uptoProxyConfig.proxyAddress,
      abi: uptoProxyConfig.proxyABI,
      functionName: "settle",
      args: buildUptoPermit2SettleArgs(permit2Payload, settlementAmount, facilitatorAddress),
      dataSuffix
    });
    const response = await waitAndReturnSettleResponse(signer, tx, payload, payer);
    return { ...response, amount: settlementAmount.toString() };
  } catch (error) {
    return mapSettleError(error, payload, payer);
  }
}

// src/upto/facilitator/scheme.ts
var UptoEvmScheme = class {
  /**
   * Creates a new UptoEvmScheme facilitator instance.
   *
   * @param signer - The EVM signer for facilitator operations
   */
  constructor(signer) {
    this.signer = signer;
    this.scheme = "upto";
    this.caipFamily = "eip155:*";
  }
  /**
   * Returns extra metadata required by the upto scheme, including the facilitator address.
   *
   * @param _ - The network identifier (unused)
   * @returns Object with facilitatorAddress, or undefined if no signer addresses are available
   */
  getExtra(_) {
    const addresses = this.signer.getAddresses();
    if (addresses.length === 0) {
      return void 0;
    }
    return { facilitatorAddress: addresses[Math.floor(Math.random() * addresses.length)] };
  }
  /**
   * Returns the list of facilitator signer addresses for the upto scheme.
   *
   * @param _ - The network identifier (unused)
   * @returns Array of facilitator signer addresses
   */
  getSigners(_) {
    return [...this.signer.getAddresses()];
  }
  /**
   * Verifies an upto Permit2 payment payload against the given requirements.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements to verify against
   * @param context - Optional facilitator context
   * @param _ - Payment required extensions (unused; reserved for interface parity)
   * @returns Promise resolving to a verification response
   */
  async verify(payload, requirements, context, _) {
    const rawPayload = payload.payload;
    if (!isUptoPermit2Payload(rawPayload)) {
      return { isValid: false, invalidReason: "unsupported_payload_type", payer: "" };
    }
    return verifyUptoPermit2(
      this.signer,
      payload,
      requirements,
      rawPayload,
      context
    );
  }
  /**
   * Settles an upto Permit2 payment on-chain.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @param context - Optional facilitator context
   * @returns Promise resolving to a settlement response
   */
  async settle(payload, requirements, context) {
    const rawPayload = payload.payload;
    if (!isUptoPermit2Payload(rawPayload)) {
      return {
        success: false,
        network: payload.accepted.network,
        transaction: "",
        errorReason: "unsupported_payload_type",
        payer: ""
      };
    }
    return settleUptoPermit2(
      this.signer,
      payload,
      requirements,
      rawPayload,
      context
    );
  }
};
export {
  UptoEvmScheme
};
